<!DOCTYPE html>
<html lang="en"><head>
<script src="deg_slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="deg_slides_files/libs/quarto-html/tabby.min.js"></script>
<script src="deg_slides_files/libs/quarto-html/popper.min.js"></script>
<script src="deg_slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="deg_slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="deg_slides_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="deg_slides_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="deg_slides_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.4.555">

  <meta name="author" content="Dr Stevie Pederson">
  <meta name="dcterms.date" content="2024-07-11">
  <title>Differential Gene Expression</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="deg_slides_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="deg_slides_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="deg_slides_files/libs/revealjs/dist/theme/quarto.css">
  <link href="deg_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="deg_slides_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="deg_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="deg_slides_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#3d3d40" data-background-image="assets/bodl_logo_white_background.jpg" data-background-opacity="0.3" data-background-size="90%" class="quarto-title-block center">
  <h1 class="title">Differential Gene Expression</h1>
  <p class="subtitle">RAdelaide 2024</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Dr Stevie Pederson 
</div>
        <p class="quarto-title-affiliation">
            </p><div class="line-block">Black Ochre Data Labs<br>
            Telethon Kids Institute</div>
          <p></p>
    </div>
</div>

  <p class="date">July 11, 2024</p>
</section>
<section>
<section id="differential-gene-expression" class="title-slide slide level1 center" data-background-color="#3c3c44" data-background-image="https://raw.githubusercontent.com/Bioconductor/BiocStickers/devel/edgeR/edgeR.png" data-background-size="30%" data-background-opacity="0.4" data-background-position="80% 50%">
<h1>Differential Gene Expression</h1>

</section>
<section id="differential-gene-expression-1" class="slide level2">
<h2>Differential Gene Expression</h2>
<ul>
<li>The most common question: <br><em>Which genes change expression levels in response to a treatment?</em></li>
</ul>
<div class="fragment">
<ul>
<li>Need to perform a statistical test on each gene
<ul>
<li>Select genes-of-interest using some criteria</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Dealing with count data <span class="math inline">\(\implies\)</span> can’t be Normally distributed</li>
<li>Generally assumed to have a Negative Binomial distribution
<ul>
<li>Essentially a Poisson distribution with additional variability</li>
</ul></li>
</ul>
</div>
</section>
<section id="todays-data" class="slide level2">
<h2>Today’s Data</h2>
<ul>
<li>Data obtained from Gene Expression Omnibus <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171742">GSE171742</a></li>
<li>Stem cell derived cardiomyocytes with 4 treatment groups <span class="citation" data-cites="Liu2023-qu">(<a href="#/references" role="doc-biblioref" onclick="">Liu et al. 2023</a>)</span></li>
<li>All transfected using lentiviral vectors</li>
</ul>
<div class="fragment">
<ol type="1">
<li>Control</li>
<li>Nsp6 over-expression</li>
<li>Nsp8 over-expression</li>
<li>M over-expression</li>
</ol>
</div>
</section>
<section id="todays-data-1" class="slide level2">
<h2>Today’s Data</h2>

<img data-src="assets/Liu_paper.png" class="quarto-figure quarto-figure-left r-stretch"></section>
<section id="todays-data-2" class="slide level2">
<h2>Today’s Data</h2>
<ul>
<li>Well documented <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM5233239">metadata</a>! 🤓</li>
<li>We know exact versions of all tools and annotations</li>
</ul>
<p><img data-src="assets/GEO_description.png" class="absolute" style="left: 0px; bottom: 100px; width: 1000px; "></p>
</section>
<section id="loading-packages" class="slide level2">
<h2>Loading Packages</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb1-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb1-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb1-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb1-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyranges)</span>
<span id="cb1-7"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-8"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb1-9"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb1-10"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb1-11"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb1-12"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="getting-annotations" class="slide level2">
<h2>Getting Annotations</h2>
<ul>
<li>Which one shall we choose?</li>
</ul>
<div style="font-size: 90%;">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="" aria-hidden="true" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb2-2"><a href="" aria-hidden="true" tabindex="-1"></a>ah <span class="sc">%&gt;%</span> <span class="fu">subset</span>(dataprovider <span class="sc">==</span> <span class="st">"Gencode"</span> <span class="sc">&amp;</span> genome <span class="sc">==</span> <span class="st">"GRCh38"</span>) <span class="sc">%&gt;%</span> <span class="fu">query</span>(<span class="st">"v31"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AnnotationHub with 9 records
# snapshotDate(): 2024-04-30
# $dataprovider: Gencode
# $species: Homo sapiens
# $rdataclass: GRanges
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., 'object[["AH75118"]]' 

            title                                                    
  AH75118 | gencode.v31.2wayconspseudos.gff3.gz                      
  AH75119 | gencode.v31.annotation.gff3.gz                           
  AH75120 | gencode.v31.basic.annotation.gff3.gz                     
  AH75121 | gencode.v31.chr_patch_hapl_scaff.annotation.gff3.gz      
  AH75122 | gencode.v31.chr_patch_hapl_scaff.basic.annotation.gff3.gz
  AH75123 | gencode.v31.long_noncoding_RNAs.gff3.gz                  
  AH75124 | gencode.v31.polyAs.gff3.gz                               
  AH75125 | gencode.v31.primary_assembly.annotation.gff3.gz          
  AH75126 | gencode.v31.tRNAs.gff3.gz                                </code></pre>
</div>
</div>
</div>
</section>
<section id="getting-annotations-1" class="slide level2">
<h2>Getting Annotations</h2>
<div style="font-size: 95%;">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="" aria-hidden="true" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> ah[[<span class="st">"AH75121"</span>]] <span class="co"># This will take several minutes</span></span>
<span id="cb4-2"><a href="" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> gtf <span class="sc">%&gt;%</span> </span>
<span id="cb4-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(type <span class="sc">==</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb4-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="st">"gene"</span>))</span>
<span id="cb4-5"><a href="" aria-hidden="true" tabindex="-1"></a>genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 66738 ranges and 3 metadata columns:
            seqnames        ranges strand |           gene_id              gene_type   gene_name
               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;character&gt;            &lt;character&gt; &lt;character&gt;
      [1]       chr1   11869-14409      + | ENSG00000223972.5 transcribed_unproces..     DDX11L1
      [2]       chr1   14404-29570      - | ENSG00000227232.5 unprocessed_pseudogene      WASH7P
      [3]       chr1   17369-17436      - | ENSG00000278267.1                  miRNA   MIR6859-1
      [4]       chr1   29554-31109      + | ENSG00000243485.5                 lncRNA MIR1302-2HG
      [5]       chr1   30366-30503      + | ENSG00000284332.1                  miRNA   MIR1302-2
      ...        ...           ...    ... .               ...                    ...         ...
  [66734] KI270734.1   72411-74814      + | ENSG00000276017.1         protein_coding  AC007325.1
  [66735] KI270734.1 131494-137392      + | ENSG00000278817.1         protein_coding  AC007325.4
  [66736] KI270734.1 138082-161852      - | ENSG00000277196.4         protein_coding  AC007325.2
  [66737] KI270744.1   51009-51114      - | ENSG00000278625.1                  snRNA     RF00026
  [66738] KI270750.1 148668-148843      + | ENSG00000277374.1                  snRNA     RF00003
  -------
  seqinfo: 407 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</div>
</section>
<section id="loading-counts" class="slide level2">
<h2>Loading Counts</h2>
<ul>
<li>Data is exactly as produced by <code>featureCounts</code> <span class="citation" data-cites="Liao2014-gy">(<a href="#/references" role="doc-biblioref" onclick="">Liao, Smyth, and Shi 2014</a>)</span>
<ul>
<li>Tab-delimited but with some weird columns</li>
</ul></li>
<li>Let’s have a sneak preview</li>
</ul>
<div class="fragment">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">read_tsv</span>(<span class="st">"data/GSE171742_counts.out.gz"</span>, <span class="at">n_max =</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 18
  Geneid      Chr   Start End   Strand Length `1_control` `6_control` `11_control` `2_Nsp6` `7_Nsp6`
  &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
1 ENSG000000… chrX… 1006… 1006… -;-;-…   4535        5211        6192         4542     4328     5283
2 ENSG000000… chrX… 1005… 1005… +;+;+…   1476          65         178           96       50       86
3 ENSG000000… chr2… 5093… 5093… -;-;-…   1207        1454        1889         1329     1472     1684
4 ENSG000000… chr1… 1698… 1698… -;-;-…   6883         628         548          522      560      636
5 ENSG000000… chr1… 1696… 1696… +;+;+…   5970         191         241          150      146      254
6 ENSG000000… chr1… 2761… 2761… -;-;-…   3382           0           0            0        0        0
# ℹ 7 more variables: `12_Nsp6` &lt;dbl&gt;, `3_Nsp8` &lt;dbl&gt;, `8_Nsp8` &lt;dbl&gt;, `13_Nsp8` &lt;dbl&gt;,
#   `4_M` &lt;dbl&gt;, `9_M` &lt;dbl&gt;, `14_M` &lt;dbl&gt;</code></pre>
</div>
</div>
</div>
<div class="fragment">
<p>Which columns do we need? How would you parse them?</p>
</div>
</section>
<section id="loading-counts-1" class="slide level2">
<h2>Loading Counts</h2>
<ul>
<li>The best form for counts is as a matrix</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/GSE171742_counts.out.gz"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(Geneid, <span class="fu">contains</span>(<span class="st">"_"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb8-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb8-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"Geneid"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb8-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> num [1:66738, 1:12] 5211 65 1454 628 191 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:66738] "ENSG00000000003.14" "ENSG00000000005.6" "ENSG00000000419.12" "ENSG00000000457.14" ...
  ..$ : chr [1:12] "1_control" "6_control" "11_control" "2_Nsp6" ...</code></pre>
</div>
</div>
</section>
<section id="checking-annotations" class="slide level2">
<h2>Checking Annotations</h2>
<ul>
<li>Compare these with the annotations we have</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">setdiff</span>(<span class="fu">rownames</span>(counts), genes<span class="sc">$</span>gene_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ENSG00000002586.20_PAR_Y" "ENSG00000124333.16_PAR_Y" "ENSG00000124334.17_PAR_Y"
 [4] "ENSG00000167393.17_PAR_Y" "ENSG00000168939.11_PAR_Y" "ENSG00000169084.14_PAR_Y"
 [7] "ENSG00000169093.16_PAR_Y" "ENSG00000169100.14_PAR_Y" "ENSG00000178605.13_PAR_Y"
[10] "ENSG00000182162.11_PAR_Y" "ENSG00000182378.14_PAR_Y" "ENSG00000182484.15_PAR_Y"
[13] "ENSG00000185203.12_PAR_Y" "ENSG00000185291.11_PAR_Y" "ENSG00000185960.14_PAR_Y"
[16] "ENSG00000196433.13_PAR_Y" "ENSG00000197976.12_PAR_Y" "ENSG00000198223.16_PAR_Y"
[19] "ENSG00000205755.11_PAR_Y" "ENSG00000214717.12_PAR_Y" "ENSG00000223274.6_PAR_Y" 
[22] "ENSG00000223484.7_PAR_Y"  "ENSG00000223511.7_PAR_Y"  "ENSG00000223571.6_PAR_Y" 
[25] "ENSG00000223773.7_PAR_Y"  "ENSG00000225661.7_PAR_Y"  "ENSG00000226179.6_PAR_Y" 
[28] "ENSG00000227159.8_PAR_Y"  "ENSG00000228410.6_PAR_Y"  "ENSG00000228572.7_PAR_Y" 
[31] "ENSG00000229232.6_PAR_Y"  "ENSG00000230542.6_PAR_Y"  "ENSG00000234622.6_PAR_Y" 
[34] "ENSG00000234958.6_PAR_Y"  "ENSG00000236017.8_PAR_Y"  "ENSG00000236871.7_PAR_Y" 
[37] "ENSG00000237040.6_PAR_Y"  "ENSG00000237531.6_PAR_Y"  "ENSG00000237801.6_PAR_Y" 
[40] "ENSG00000265658.6_PAR_Y"  "ENSG00000270726.6_PAR_Y"  "ENSG00000275287.5_PAR_Y" 
[43] "ENSG00000277120.5_PAR_Y"  "ENSG00000280767.3_PAR_Y"  "ENSG00000281849.3_PAR_Y" </code></pre>
</div>
</div>
<aside class="notes">
<ul>
<li>If we’re doing everything, we don’t need to check these</li>
<li>Only for public data. Annotations can be difficult to untangle</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<ul>
<li>Should we remove the suffixes?</li>
</ul>
</section>
<section id="checking-annotations-1" class="slide level2">
<h2>Checking Annotations</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="" aria-hidden="true" tabindex="-1"></a>genes <span class="sc">%&gt;%</span> <span class="fu">subset</span>(gene_id <span class="sc">==</span> <span class="st">"ENSG00000002586.20"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 2 ranges and 3 metadata columns:
      seqnames          ranges strand |            gene_id      gene_type   gene_name
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt;    &lt;character&gt; &lt;character&gt;
  [1]     chrX 2691187-2741309      + | ENSG00000002586.20 protein_coding        CD99
  [2]     chrY 2691187-2741309      + | ENSG00000002586.20 protein_coding        CD99
  -------
  seqinfo: 407 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<ul>
<li>How can we resolve this?</li>
</ul>
</section>
<section id="checking-annotations-2" class="slide level2">
<h2>Checking Annotations</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> genes <span class="sc">%&gt;%</span> </span>
<span id="cb14-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb14-3"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene_id =</span> <span class="fu">case_when</span>(</span>
<span id="cb14-4"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="fu">duplicated</span>(gene_id) <span class="sc">~</span> <span class="fu">paste0</span>(gene_id, <span class="st">"_PAR_Y"</span>),</span>
<span id="cb14-5"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> gene_id</span>
<span id="cb14-6"><a href="" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb14-7"><a href="" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb14-8"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(genes, <span class="fu">duplicated</span>(gene_id))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 0 ranges and 3 metadata columns:
   seqnames    ranges strand |     gene_id   gene_type   gene_name
      &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;
  -------
  seqinfo: 407 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section></section>
<section>
<section id="dge-analysis" class="title-slide slide level1 center" data-background-color="#3c3c44" data-background-image="https://raw.githubusercontent.com/Bioconductor/BiocStickers/devel/edgeR/edgeR.png" data-background-size="30%" data-background-opacity="0.4" data-background-position="80% 50%">
<h1>DGE Analysis</h1>

</section>
<section id="basic-workflow" class="slide level2">
<h2>Basic Workflow</h2>
<ol type="1">
<li>Remove low-expressed &amp; undetectable genes</li>
<li>Normalise the data</li>
<li>Estimate dispersions</li>
<li>Perform Statistical Tests</li>
<li>Enrichment Testing</li>
</ol>
<div class="fragment">
<ul>
<li>Careful checks at every step to inform decisions</li>
</ul>
</div>
</section>
<section id="dgelist-objects" class="slide level2">
<h2>DGEList Objects</h2>
<ul>
<li>We’ll use <code>edgeR</code> for DGE analysis <span class="citation" data-cites="edgeR2010">(<a href="#/references" role="doc-biblioref" onclick="">Robinson, McCarthy, and Smyth 2010</a>)</span></li>
<li>Counts, sample metadata and gene annotations stored in a single <code>S4</code> object
<ul>
<li><code>DGEList</code>: <em>Digital</em> Gene Expression List</li>
<li>Gene &amp; metadata elements need to be <code>data.frame</code> objects</li>
<li>Will be coerced if able . . .</li>
</ul></li>
<li><code>DESeq2</code> is a common alternative <span class="citation" data-cites="DESeq22014">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span></li>
<li>Uses an extension of <code>SummarizedExperiment</code> objects</li>
</ul>
<aside class="notes">
<ul>
<li>Both methods are gold-standard but neither perfect for every dataset</li>
<li>Both Wolfgang Huber &amp; Gordon Smyth are exceptional statisticians</li>
<li>Will try get to DESeq2 later in the session</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="dgelist-objects-1" class="slide level2">
<h2>DGEList Objects</h2>
<ul>
<li>How can we get sample metadata?</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">id =</span> <span class="fu">colnames</span>(counts)) <span class="sc">%&gt;%</span> </span>
<span id="cb16-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(id, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"replicate"</span>, <span class="st">"treatment"</span>), <span class="at">remove =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb16-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb16-4"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">treatment =</span> <span class="fu">as.factor</span>(treatment),</span>
<span id="cb16-5"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">group =</span> <span class="fu">as.integer</span>(treatment)</span>
<span id="cb16-6"><a href="" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb16-7"><a href="" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 4
   id         replicate treatment group
   &lt;chr&gt;      &lt;chr&gt;     &lt;fct&gt;     &lt;int&gt;
 1 1_control  1         control       1
 2 6_control  6         control       1
 3 11_control 11        control       1
 4 2_Nsp6     2         Nsp6          3
 5 7_Nsp6     7         Nsp6          3
 6 12_Nsp6    12        Nsp6          3
 7 3_Nsp8     3         Nsp8          4
 8 8_Nsp8     8         Nsp8          4
 9 13_Nsp8    13        Nsp8          4
10 4_M        4         M             2
11 9_M        9         M             2
12 14_M       14        M             2</code></pre>
</div>
</div>
<aside class="notes">
<p>Will be in the same order as <code>counts</code></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="dgelist-objects-2" class="slide level2">
<h2>DGEList Objects</h2>
<ul>
<li>Annotations appear to match the counts
<ul>
<li>Counts for 66,738 genes</li>
<li>Annotations for 66,738 genes</li>
</ul></li>
<li>But will be in a different order
<ul>
<li>Counts in order of name</li>
<li>Annotations in genomic order</li>
</ul></li>
</ul>
</section>
<section id="dgelist-objects-3" class="slide level2">
<h2>DGEList Objects</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="" aria-hidden="true" tabindex="-1"></a>genes <span class="sc">%&gt;%</span> <span class="fu">setNames</span>(.<span class="sc">$</span>gene_id) <span class="sc">%&gt;%</span> .[<span class="fu">rownames</span>(counts)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 66738 ranges and 3 metadata columns:
                     seqnames              ranges strand |            gene_id      gene_type
                        &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt;    &lt;character&gt;
  ENSG00000000003.14     chrX 100627109-100639991      - | ENSG00000000003.14 protein_coding
   ENSG00000000005.6     chrX 100584936-100599885      + |  ENSG00000000005.6 protein_coding
  ENSG00000000419.12    chr20   50934867-50958555      - | ENSG00000000419.12 protein_coding
  ENSG00000000457.14     chr1 169849631-169894267      - | ENSG00000000457.14 protein_coding
  ENSG00000000460.17     chr1 169662007-169854080      + | ENSG00000000460.17 protein_coding
                 ...      ...                 ...    ... .                ...            ...
   ENSG00000288107.1    chr10 133565797-133629507      + |  ENSG00000288107.1         lncRNA
   ENSG00000288108.1     chr5     1931058-1933985      - |  ENSG00000288108.1         lncRNA
   ENSG00000288109.1    chr17   67658015-67659465      - |  ENSG00000288109.1         lncRNA
   ENSG00000288110.1     chr8     4496392-4503392      + |  ENSG00000288110.1         lncRNA
   ENSG00000288111.1     chr3 130181346-130188814      + |  ENSG00000288111.1         lncRNA
                       gene_name
                     &lt;character&gt;
  ENSG00000000003.14      TSPAN6
   ENSG00000000005.6        TNMD
  ENSG00000000419.12        DPM1
  ENSG00000000457.14       SCYL3
  ENSG00000000460.17    C1orf112
                 ...         ...
   ENSG00000288107.1  AL731769.2
   ENSG00000288108.1  AC126768.3
   ENSG00000288109.1  AC079331.3
   ENSG00000288110.1  AC010941.1
   ENSG00000288111.1  AC130888.1
  -------
  seqinfo: 407 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="dgelist-objects-4" class="slide level2">
<h2>DGEList Objects</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="" aria-hidden="true" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(</span>
<span id="cb20-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> counts,</span>
<span id="cb20-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">samples =</span> <span class="fu">tibble</span>(<span class="at">id =</span> <span class="fu">colnames</span>(counts)) <span class="sc">%&gt;%</span> <span class="fu">left_join</span>(samples),</span>
<span id="cb20-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">genes =</span> genes <span class="sc">%&gt;%</span> </span>
<span id="cb20-5"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setNames</span>(.<span class="sc">$</span>gene_id) <span class="sc">%&gt;%</span> </span>
<span id="cb20-6"><a href="" aria-hidden="true" tabindex="-1"></a>    .[<span class="fu">rownames</span>(counts)] <span class="sc">%&gt;%</span> </span>
<span id="cb20-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(<span class="at">row.names =</span> <span class="fu">names</span>(.))</span>
<span id="cb20-8"><a href="" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb20-9"><a href="" aria-hidden="true" tabindex="-1"></a>dge</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class "DGEList"
$counts
                   1_control 6_control 11_control 2_Nsp6 7_Nsp6 12_Nsp6 3_Nsp8 8_Nsp8 13_Nsp8  4_M
ENSG00000000003.14      5211      6192       4542   4328   5283    4336   4330   5208    4635 4743
ENSG00000000005.6         65       178         96     50     86      83     67     92      80   51
ENSG00000000419.12      1454      1889       1329   1472   1684    1368   1666   2042    1707 1822
ENSG00000000457.14       628       548        522    560    636     492    530    500     498  622
ENSG00000000460.17       191       241        150    146    254     164    161    179     183  173
                    9_M 14_M
ENSG00000000003.14 5498 5172
ENSG00000000005.6    70   64
ENSG00000000419.12 2293 1987
ENSG00000000457.14  513  560
ENSG00000000460.17  256  228
66733 more rows ...

$samples
           group lib.size norm.factors         id replicate treatment
1_control      1 38633196            1  1_control         1   control
6_control      1 48184412            1  6_control         6   control
11_control     1 34566631            1 11_control        11   control
2_Nsp6         3 33088151            1     2_Nsp6         2      Nsp6
7_Nsp6         3 41621703            1     7_Nsp6         7      Nsp6
7 more rows ...

$genes
                   seqnames     start       end  width strand            gene_id      gene_type
ENSG00000000003.14     chrX 100627109 100639991  12883      - ENSG00000000003.14 protein_coding
ENSG00000000005.6      chrX 100584936 100599885  14950      +  ENSG00000000005.6 protein_coding
ENSG00000000419.12    chr20  50934867  50958555  23689      - ENSG00000000419.12 protein_coding
ENSG00000000457.14     chr1 169849631 169894267  44637      - ENSG00000000457.14 protein_coding
ENSG00000000460.17     chr1 169662007 169854080 192074      + ENSG00000000460.17 protein_coding
                   gene_name
ENSG00000000003.14    TSPAN6
ENSG00000000005.6       TNMD
ENSG00000000419.12      DPM1
ENSG00000000457.14     SCYL3
ENSG00000000460.17  C1orf112
66733 more rows ...</code></pre>
</div>
</div>
</section>
<section id="dgelist-objects-5" class="slide level2">
<h2>DGEList Objects</h2>
<ul>
<li>Key elements have different but related dimensions</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 66738    12</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge<span class="sc">$</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 66738    12</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge<span class="sc">$</span>samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12  6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(dge<span class="sc">$</span>genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 66738     8</code></pre>
</div>
</div>
<ul>
<li><code>samples</code> is column metadata <span class="math inline">\(\implies\)</span> <code>genes</code> is row metadata</li>
</ul>
</section>
<section id="library-sizes" class="slide level2">
<h2>Library Sizes</h2>
<ul>
<li>The very first step with counts <span class="math inline">\(\implies\)</span> library sizes</li>
<li>This is the total number of reads (i.e.&nbsp;RNA fragments) assigned to genes</li>
<li>Ideally &gt;20m reads per sample</li>
<li>Those below 10m reads are generally considered as problematic
<ul>
<li>Context dependent</li>
</ul></li>
</ul>
</section>
<section id="library-sizes-1" class="slide level2">
<h2>Library Sizes</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>samples <span class="sc">%&gt;%</span> </span>
<span id="cb30-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(replicate, lib.size <span class="sc">/</span> <span class="fl">1e6</span>, <span class="at">fill =</span> treatment)) <span class="sc">+</span></span>
<span id="cb30-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb30-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>treatment, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb30-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Sample"</span>, <span class="at">y =</span> <span class="st">"Library Size (millions)"</span>, <span class="at">fill =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb30-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>))) <span class="sc">+</span></span>
<span id="cb30-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-lib-size-1.png">Click to show figure</a></p>
</div>
</section>
<section id="library-sizes-2" class="slide level2">
<h2>Library Sizes</h2>
<ul>
<li>We can get fancy and add the mean value</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>samples <span class="sc">%&gt;%</span> </span>
<span id="cb31-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(replicate, lib.size <span class="sc">/</span> <span class="fl">1e6</span>, <span class="at">fill =</span> treatment)) <span class="sc">+</span></span>
<span id="cb31-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb31-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(</span>
<span id="cb31-5"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">yintercept =</span> lib.size),</span>
<span id="cb31-6"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> . <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">lib.size =</span> <span class="fu">mean</span>(lib.size <span class="sc">/</span> <span class="fl">1e6</span>)),</span>
<span id="cb31-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">colour =</span> <span class="st">"grey30"</span></span>
<span id="cb31-8"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb31-9"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>treatment, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb31-10"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Sample"</span>, <span class="at">y =</span> <span class="st">"Library Size (millions)"</span>, <span class="at">fill =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb31-11"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>))) <span class="sc">+</span></span>
<span id="cb31-12"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/lib-size-with-mean-1.png">Click to show figure</a></p>
</div>
<aside class="notes">
<p>These look like really good libraries so far</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="detected-vs-undetected-genes" class="slide level2">
<h2>Detected Vs Undetected Genes</h2>
<ul>
<li>By adding counts across samples and seeing how many are zero<br><span class="math inline">\(\implies\)</span> non-detectable genes</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="" aria-hidden="true" tabindex="-1"></a>dge<span class="sc">$</span>counts <span class="sc">%&gt;%</span> </span>
<span id="cb32-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowSums</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb32-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">equals</span>(<span class="dv">0</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb32-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 27211</code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>Genes with low counts can also be removed
<ul>
<li>These end up being highly variable</li>
<li>Uninformative statistically</li>
</ul></li>
</ul>
</div>
</section>
<section id="detected-vs-undetected-genes-1" class="slide level2">
<h2>Detected Vs Undetected Genes</h2>
<ul>
<li>No set method for considering a gene as <em>too low to bother with</em></li>
<li>Can use raw counts below some value</li>
<li>Counts within a group above some value <span class="math inline">\(\implies\)</span> detected within a treatment</li>
</ul>
<div class="fragment">
<ul>
<li>Counts per million reads (CPM) is a common measure used in abundance analysis</li>
<li>Take the library sizes / 1e6 &amp; divide counts by this value <span class="math inline">\(\implies\)</span> CPM</li>
<li>Can also use logCPM
<ul>
<li>Adds a non-zero value to every gene to avoid zeroes</li>
</ul></li>
</ul>
</div>
</section>
<section id="detected-vs-undetected-genes-2" class="slide level2">
<h2>Detected Vs Undetected Genes</h2>
<ul>
<li><code>edgeR</code> provides a function <code>filterByExpr()</code></li>
<li>Flags genes for removal or discarding based on counts</li>
<li>Set minimum CPM as <code>min.count</code> <span class="math inline">\(\times 1e6\)</span> / <code>median(lib.size)</code></li>
<li>Set minimum <em>total counts</em></li>
</ul>
<div class="fragment">
<ul>
<li>Genes are retained if CPM &gt; minCPM in <span class="math inline">\(\geq n\)</span> samples
<ul>
<li><span class="math inline">\(n\)</span> is the smallest treatment group</li>
</ul></li>
<li>Default settings may be a little generous for this dataset</li>
</ul>
</div>
</section>
<section id="detected-vs-undetected-genes-3" class="slide level2">
<h2>Detected Vs Undetected Genes</h2>
<ul>
<li>If we require &gt; 1CPM in at least 3 samples</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="" aria-hidden="true" tabindex="-1"></a>min_count <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples<span class="sc">$</span>lib.size <span class="sc">%&gt;%</span> </span>
<span id="cb34-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">median</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb34-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">divide_by</span>(<span class="fl">1e6</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb34-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>(<span class="dv">0</span>)</span>
<span id="cb34-5"><a href="" aria-hidden="true" tabindex="-1"></a>min_count</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 38</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="" aria-hidden="true" tabindex="-1"></a>min_total <span class="ot">&lt;-</span> <span class="dv">150</span></span>
<span id="cb36-2"><a href="" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, <span class="at">min.count =</span> min_count, <span class="at">min.total.count =</span> <span class="dv">150</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="detected-vs-undetected-genes-4" class="slide level2">
<h2>Detected Vs Undetected Genes</h2>
<div style="font-size: 95%;">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> dge<span class="sc">$</span>counts <span class="sc">%&gt;%</span> </span>
<span id="cb37-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cpm</span>(<span class="at">log =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"gene_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="fu">all_of</span>(<span class="st">"gene_id"</span>), <span class="at">names_to =</span> <span class="st">"id"</span>, <span class="at">values_to =</span> <span class="st">"logCPM"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(dge<span class="sc">$</span>samples) <span class="sc">%&gt;%</span> </span>
<span id="cb37-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(logCPM, <span class="fu">after_stat</span>(density), <span class="at">group =</span> id)) <span class="sc">+</span></span>
<span id="cb37-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">colour =</span> treatment)) <span class="sc">+</span></span>
<span id="cb37-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="fu">glue</span>(<span class="st">"Before Filtering: {comma(length(keep))} genes"</span>)) <span class="sc">+</span></span>
<span id="cb37-9"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Density"</span>, <span class="at">colour =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb37-10"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb37-11"><a href="" aria-hidden="true" tabindex="-1"></a>B <span class="ot">&lt;-</span> dge<span class="sc">$</span>counts[keep,] <span class="sc">%&gt;%</span> </span>
<span id="cb37-12"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cpm</span>(<span class="at">log =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-13"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"gene_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-14"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="fu">all_of</span>(<span class="st">"gene_id"</span>), <span class="at">names_to =</span> <span class="st">"id"</span>, <span class="at">values_to =</span> <span class="st">"logCPM"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-15"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(dge<span class="sc">$</span>samples) <span class="sc">%&gt;%</span> </span>
<span id="cb37-16"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(logCPM, <span class="fu">after_stat</span>(density), <span class="at">group =</span> id)) <span class="sc">+</span></span>
<span id="cb37-17"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">colour =</span> treatment)) <span class="sc">+</span></span>
<span id="cb37-18"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="fu">glue</span>(<span class="st">"After Filtering: {comma(sum(keep))} genes"</span>)) <span class="sc">+</span></span>
<span id="cb37-19"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Density"</span>, <span class="at">colour =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb37-20"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb37-21"><a href="" aria-hidden="true" tabindex="-1"></a>A <span class="sc">+</span> B <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>, <span class="at">axes =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-densities-1.png">Click to show figure</a></p>
</div>
</section>
<section id="detected-vs-undetected-genes-5" class="slide level2">
<h2>Detected Vs Undetected Genes</h2>
<ul>
<li>Create new object once satisfied with the filtering strategy</li>
<li>Passing the logical vector will subset <code>counts</code> &amp; <code>genes</code> correctly
<ul>
<li>The original library sizes are retained</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="" aria-hidden="true" tabindex="-1"></a>dge_filter <span class="ot">&lt;-</span> dge[keep,]</span>
<span id="cb38-2"><a href="" aria-hidden="true" tabindex="-1"></a>dge_filter</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class "DGEList"
$counts
                   1_control 6_control 11_control 2_Nsp6 7_Nsp6 12_Nsp6 3_Nsp8 8_Nsp8 13_Nsp8  4_M
ENSG00000000003.14      5211      6192       4542   4328   5283    4336   4330   5208    4635 4743
ENSG00000000005.6         65       178         96     50     86      83     67     92      80   51
ENSG00000000419.12      1454      1889       1329   1472   1684    1368   1666   2042    1707 1822
ENSG00000000457.14       628       548        522    560    636     492    530    500     498  622
ENSG00000000460.17       191       241        150    146    254     164    161    179     183  173
                    9_M 14_M
ENSG00000000003.14 5498 5172
ENSG00000000005.6    70   64
ENSG00000000419.12 2293 1987
ENSG00000000457.14  513  560
ENSG00000000460.17  256  228
13891 more rows ...

$samples
           group lib.size norm.factors         id replicate treatment
1_control      1 38633196            1  1_control         1   control
6_control      1 48184412            1  6_control         6   control
11_control     1 34566631            1 11_control        11   control
2_Nsp6         3 33088151            1     2_Nsp6         2      Nsp6
7_Nsp6         3 41621703            1     7_Nsp6         7      Nsp6
7 more rows ...

$genes
                   seqnames     start       end  width strand            gene_id      gene_type
ENSG00000000003.14     chrX 100627109 100639991  12883      - ENSG00000000003.14 protein_coding
ENSG00000000005.6      chrX 100584936 100599885  14950      +  ENSG00000000005.6 protein_coding
ENSG00000000419.12    chr20  50934867  50958555  23689      - ENSG00000000419.12 protein_coding
ENSG00000000457.14     chr1 169849631 169894267  44637      - ENSG00000000457.14 protein_coding
ENSG00000000460.17     chr1 169662007 169854080 192074      + ENSG00000000460.17 protein_coding
                   gene_name
ENSG00000000003.14    TSPAN6
ENSG00000000005.6       TNMD
ENSG00000000419.12      DPM1
ENSG00000000457.14     SCYL3
ENSG00000000460.17  C1orf112
13891 more rows ...</code></pre>
</div>
</div>
</section>
<section id="normalisation" class="slide level2">
<h2>Normalisation</h2>
<ul>
<li>Normalisation is a very common step in transcriptomic analysis</li>
<li>Helps to account for differences in library composition
<ul>
<li>e.g.&nbsp;are some libraries dominated by a handful of highly abundant genes</li>
</ul></li>
<li>The default in <code>edgeR</code> is TMM <span class="citation" data-cites="Robinson2010-qp">(<a href="#/references" role="doc-biblioref" onclick="">Robinson and Oshlack 2010</a>)</span></li>
<li><code>DESeq2</code> uses RLE normalisation
<ul>
<li>Very similar in principle</li>
</ul></li>
</ul>
</section>
<section id="statistical-analysis-of-counts" class="slide level2">
<h2>Statistical Analysis of Counts</h2>
<ul>
<li>Poisson models number of events per unit
<ul>
<li>e.g.&nbsp;telephone calls per minute</li>
</ul></li>
<li>The rate parameter (<span class="math inline">\(\lambda\)</span>) models this</li>
<li>mean occurrence = average rate = <span class="math inline">\(\lambda\)</span></li>
<li>Variance is fixed at <span class="math inline">\(\lambda\)</span>
<ul>
<li>If variance <span class="math inline">\(\neq \lambda \implies\)</span> not Poisson data</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>The basic model for <code>edgeR</code> is a Negative Binomial Model
<ul>
<li>NB: mean = <span class="math inline">\(\lambda\)</span>; variance = <span class="math inline">\(\lambda + r\)</span></li>
</ul></li>
<li>Multiple strategies for obtaining estimates</li>
</ul>
</div>
</section>
<section id="statistical-analysis-of-counts-1" class="slide level2">
<h2>Statistical Analysis of Counts</h2>
<ul>
<li>Counts / gene (i.e.&nbsp;<span class="math inline">\(y_{gi}\)</span>) are a function of library size (<span class="math inline">\(L_i\)</span>)</li>
<li>Gene length is fixed across samples <span class="math inline">\(\implies\)</span> library size is not</li>
<li>Dividing by library size is a way of normalising counts</li>
</ul>
<p><span class="math display">\[
\log (y_{gi} / L_i) = \beta_0 + \beta_1 + ...
\]</span></p>
</section>
<section id="normalisation-1" class="slide level2">
<h2>Normalisation</h2>
<ul>
<li>Calculation of normalisation factors (<span class="math inline">\(Y_i\)</span>) per sample
<ul>
<li>The idealised normalisation factor is <span class="math inline">\(Y_i = 1\)</span></li>
</ul></li>
<li>Can use scaled library sizes <span class="math inline">\(L^*_i = Y_i \times L_i\)</span></li>
</ul>
<p><span class="math display">\[
\log (y_{gi} / L^*_i) = \beta_0 + \beta_1 + ...
\]</span></p>
</section>
<section id="tmm-normalisation" class="slide level2">
<h2>TMM Normalisation</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="" aria-hidden="true" tabindex="-1"></a>dge_filter <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge_filter)</span>
<span id="cb40-2"><a href="" aria-hidden="true" tabindex="-1"></a>dge_filter<span class="sc">$</span>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           group lib.size norm.factors         id replicate treatment
1_control      1 38633196    0.9943180  1_control         1   control
6_control      1 48184412    0.9770121  6_control         6   control
11_control     1 34566631    0.9883140 11_control        11   control
2_Nsp6         3 33088151    1.0157571     2_Nsp6         2      Nsp6
7_Nsp6         3 41621703    0.9963827     7_Nsp6         7      Nsp6
12_Nsp6        3 32724713    1.0060007    12_Nsp6        12      Nsp6
3_Nsp8         4 33502022    1.0076816     3_Nsp8         3      Nsp8
8_Nsp8         4 41364663    0.9977052     8_Nsp8         8      Nsp8
13_Nsp8        4 36288023    1.0027832    13_Nsp8        13      Nsp8
4_M            2 36851725    1.0081808        4_M         4         M
9_M            2 41546877    0.9987653        9_M         9         M
14_M           2 43587502    1.0076991       14_M        14         M</code></pre>
</div>
</div>
<aside class="notes">
<ul>
<li>The clustering around 1 suggests very similar libraries</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="pca-analysis" class="slide level2">
<h2>PCA Analysis</h2>
<ul>
<li><code>edgeR</code> offers <code>plotMDS()</code> for quick exploration</li>
</ul>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMDS</span>(</span>
<span id="cb42-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dge_filter, </span>
<span id="cb42-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">as.integer</span>(dge_filter<span class="sc">$</span>samples<span class="sc">$</span>treatment)</span>
<span id="cb42-4"><a href="" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/plot-mds-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
<aside class="notes">
<ul>
<li>Takes top 500 most variable genes</li>
<li>Uses base graphics</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="pca-analysis-1" class="slide level2">
<h2>PCA Analysis</h2>
<ul>
<li>logCPM values are very useful for PCA
<ul>
<li>Will now be calculated using the normalisation factors</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> dge_filter <span class="sc">%&gt;%</span> </span>
<span id="cb43-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cpm</span>(<span class="at">log =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb43-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb43-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>(<span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb43-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of components:
                           PC1     PC2     PC3     PC4     PC5      PC6      PC7      PC8      PC9
Standard deviation     66.6493 50.4749 45.2310 31.7401 29.3998 25.88752 24.74707 21.51801 20.94233
Proportion of Variance  0.3197  0.1833  0.1472  0.0725  0.0622  0.04823  0.04407  0.03332  0.03156
Cumulative Proportion   0.3197  0.5030  0.6502  0.7227  0.7849  0.83316  0.87723  0.91056  0.94212
                           PC10     PC11      PC12
Standard deviation     20.27916 19.82675 2.753e-13
Proportion of Variance  0.02959  0.02829 0.000e+00
Cumulative Proportion   0.97171  1.00000 1.000e+00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="" aria-hidden="true" tabindex="-1"></a>prop_var <span class="ot">&lt;-</span> <span class="fu">summary</span>(pca)<span class="sc">$</span>importance[<span class="st">"Proportion of Variance"</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pca-analysis-2" class="slide level2">
<h2>PCA Analysis</h2>
<ul>
<li>PC1 vs PC2 shown below <span class="math inline">\(\implies\)</span> easily modified to other components</li>
</ul>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">Simple PCA</a></li><li><a href="#tabset-1-2">Enhanced Axis Labels</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="" aria-hidden="true" tabindex="-1"></a>pca <span class="sc">%&gt;%</span> </span>
<span id="cb46-2"><a href="" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb46-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">"PC"</span>, <span class="at">names_prefix =</span> <span class="st">"PC"</span>, <span class="at">values_from =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb46-4"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">id =</span> row) <span class="sc">%&gt;%</span> </span>
<span id="cb46-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(dge_filter<span class="sc">$</span>samples) <span class="sc">%&gt;%</span> </span>
<span id="cb46-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(PC1, PC2, <span class="at">colour =</span> treatment)) <span class="sc">+</span></span>
<span id="cb46-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> id), <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb46-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-pca-1.png">Click to show figure</a></p>
</div>
</div>
<div id="tabset-1-2">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="" aria-hidden="true" tabindex="-1"></a>pca <span class="sc">%&gt;%</span> </span>
<span id="cb47-2"><a href="" aria-hidden="true" tabindex="-1"></a>  broom<span class="sc">::</span><span class="fu">tidy</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb47-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">"PC"</span>, <span class="at">names_prefix =</span> <span class="st">"PC"</span>, <span class="at">values_from =</span> <span class="st">"value"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb47-4"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">id =</span> row) <span class="sc">%&gt;%</span> </span>
<span id="cb47-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(dge_filter<span class="sc">$</span>samples) <span class="sc">%&gt;%</span> </span>
<span id="cb47-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(PC1, PC2, <span class="at">colour =</span> treatment)) <span class="sc">+</span></span>
<span id="cb47-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb47-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> id), <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb47-9"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb47-10"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">glue</span>(<span class="st">"PC1 ({percent(prop_var['PC1'])} Variance)"</span>),</span>
<span id="cb47-11"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">glue</span>(<span class="st">"PC2 ({percent(prop_var['PC2'])} Variance)"</span>),</span>
<span id="cb47-12"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour =</span> <span class="st">"Treatment"</span></span>
<span id="cb47-13"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb47-14"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-pca-labels-1.png">Click to show figure</a></p>
</div>
</div>
</div>
</div>
</section>
<section id="pca-analysis-3" class="slide level2">
<h2>PCA Analysis</h2>
<ul>
<li>What do we think the above is telling us?</li>
<li>Does it look like treatment is contributing to the variance?</li>
<li>Are there any other potential sources of variability?</li>
</ul>
<aside class="notes">
<ul>
<li>Looks to me like samples 2-4 had transfection issues</li>
<li>Need to discuss with wet lab folks</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-design-matrix" class="slide level2">
<h2>The Design Matrix</h2>
<div style="font-size: 92%;">
<ul>
<li>Before proceeding need to define a <em>design matrix</em></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span>treatment, dge_filter<span class="sc">$</span>samples)</span>
<span id="cb48-2"><a href="" aria-hidden="true" tabindex="-1"></a>X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           (Intercept) treatmentM treatmentNsp6 treatmentNsp8
1_control            1          0             0             0
6_control            1          0             0             0
11_control           1          0             0             0
2_Nsp6               1          0             1             0
7_Nsp6               1          0             1             0
12_Nsp6              1          0             1             0
3_Nsp8               1          0             0             1
8_Nsp8               1          0             0             1
13_Nsp8              1          0             0             1
4_M                  1          1             0             0
9_M                  1          1             0             0
14_M                 1          1             0             0
attr(,"assign")
[1] 0 1 1 1
attr(,"contrasts")
attr(,"contrasts")$treatment
[1] "contr.treatment"</code></pre>
</div>
</div>
</div>
</section>
<section id="the-design-matrix-1" class="slide level2">
<h2>The Design Matrix</h2>
<ul>
<li>The first column will estimate average expression in control samples</li>
<li>Remaining columns estimate the difference due to treatments</li>
<li>(Almost) always modelled on the log2 scale <span class="math inline">\(\implies\)</span> logFC</li>
</ul>
</section>
<section id="estimating-dispersions" class="slide level2">
<h2>Estimating Dispersions</h2>
<ul>
<li>We can estimate the dispersion parameter across the entire dataset</li>
<li><code>edgeR</code> models consider technical noise as Poisson with <em>biological noise</em> as over-dispersion</li>
<li>This adds multiple elements to the <code>DGEList</code>
<ul>
<li>Essential for <code>edgeR</code> models</li>
</ul></li>
<li>Initial estimates are moderated using a global trend during model fitting</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="" aria-hidden="true" tabindex="-1"></a>dge_filter <span class="ot">&lt;-</span> <span class="fu">estimateDisp</span>(dge_filter, <span class="at">design =</span> X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="estimating-dispersions-1" class="slide level2">
<h2>Estimating Dispersions</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="co"># Output uses base graphics no ggplot</span></span>
<span id="cb51-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotBCV</span>(dge_filter, <span class="at">pch =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/plot-bcv-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="model-fitting" class="slide level2">
<h2>Model Fitting</h2>
<ul>
<li>The recommended approach for <code>edgeR</code> is to use <em>Quasi-Likelihood</em> fits <span class="citation" data-cites="Lund2012-xo">(<a href="#/references" role="doc-biblioref" onclick="">Lund et al. 2012</a>)</span>
<ul>
<li>Allows flexibility in over-dispersion parameters</li>
</ul></li>
<li>First we fit the model <span class="math inline">\(\implies\)</span> then perform testing</li>
<li>Using our design matrix, we will fit the model for all coefficients in <code>X</code>
<ul>
<li>Perform tests within each co-efficient</li>
<li>We’ll use the ‘M’ treatment</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">glmQLFit</span>(dge_filter, <span class="at">design =</span> X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="model-testing" class="slide level2">
<h2>Model Testing</h2>
<ul>
<li>Statistical tests use quasi-likelihood F-tests</li>
<li>Use moderate dispersions (Empirical Bayes)</li>
<li>By default, the top10 genes are returned</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="" aria-hidden="true" tabindex="-1"></a>fit <span class="sc">%&gt;%</span> <span class="fu">glmQLFTest</span>(<span class="at">coef =</span> <span class="st">"treatmentM"</span>) <span class="sc">%&gt;%</span> <span class="fu">topTags</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Coefficient:  treatmentM 
                   seqnames     start       end  width strand            gene_id      gene_type
ENSG00000065609.14     chr6  83552880  83709691 156812      - ENSG00000065609.14 protein_coding
ENSG00000133878.9      chr8  33591330  33600023   8694      -  ENSG00000133878.9 protein_coding
ENSG00000112837.17     chr6  84687351  84764598  77248      - ENSG00000112837.17 protein_coding
ENSG00000125730.16    chr19   6677704   6730562  52859      - ENSG00000125730.16 protein_coding
ENSG00000144810.16     chr3  99638475  99799226 160752      + ENSG00000144810.16 protein_coding
ENSG00000112562.18     chr6 168441151 168673445 232295      + ENSG00000112562.18 protein_coding
ENSG00000118137.9     chr11 116835751 116837950   2200      -  ENSG00000118137.9 protein_coding
ENSG00000101335.10    chr20  36541497  36551447   9951      + ENSG00000101335.10 protein_coding
ENSG00000156113.23    chr10  76869601  77638369 768769      - ENSG00000156113.23 protein_coding
ENSG00000165118.15     chr9  83938311  83956986  18676      - ENSG00000165118.15 protein_coding
                   gene_name     logFC    logCPM         F       PValue          FDR
ENSG00000065609.14    SNAP91 -4.298482 5.7855158 1244.7901 4.457146e-13 6.193650e-09
ENSG00000133878.9     DUSP26 -6.406627 0.9996001  320.3749 8.859485e-11 6.155570e-07
ENSG00000112837.17     TBX18  4.614608 1.3852329  463.1117 1.329742e-10 6.159367e-07
ENSG00000125730.16        C3  7.404154 1.9905031  222.5631 7.853570e-10 2.728330e-06
ENSG00000144810.16    COL8A1  1.897573 4.2552562  221.7729 7.221342e-09 2.006955e-05
ENSG00000112562.18     SMOC2  2.975471 2.9960180  214.3021 8.720084e-09 2.019572e-05
ENSG00000118137.9      APOA1 -1.909907 6.8931539  201.6088 1.216484e-08 2.414894e-05
ENSG00000101335.10      MYL9  2.208139 9.0902053  192.7708 1.553354e-08 2.553771e-05
ENSG00000156113.23    KCNMA1  2.906831 4.0355987  190.5689 1.653997e-08 2.553771e-05
ENSG00000165118.15   C9orf64  5.728040 1.8379059  176.1002 2.011799e-08 2.556070e-05</code></pre>
</div>
</div>
</section>
<section id="model-testing-1" class="slide level2">
<h2>Model Testing</h2>
<ul>
<li>Can easily collect all genes into a manageable object</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_M <span class="ot">&lt;-</span> fit <span class="sc">%&gt;%</span> </span>
<span id="cb55-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glmQLFTest</span>(<span class="at">coef =</span> <span class="st">"treatmentM"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTags</span>(<span class="at">n =</span> <span class="cn">Inf</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb55-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb55-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()</span>
<span id="cb55-6"><a href="" aria-hidden="true" tabindex="-1"></a>res_M</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 13,896 × 13
   seqnames     start      end  width strand gene_id gene_type gene_name logFC logCPM     F   PValue
   &lt;fct&gt;        &lt;int&gt;    &lt;int&gt;  &lt;int&gt; &lt;fct&gt;  &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;
 1 chr6      83552880   8.37e7 156812 -      ENSG00… protein_… SNAP91    -4.30   5.79 1245. 4.46e-13
 2 chr8      33591330   3.36e7   8694 -      ENSG00… protein_… DUSP26    -6.41   1.00  320. 8.86e-11
 3 chr6      84687351   8.48e7  77248 -      ENSG00… protein_… TBX18      4.61   1.39  463. 1.33e-10
 4 chr19      6677704   6.73e6  52859 -      ENSG00… protein_… C3         7.40   1.99  223. 7.85e-10
 5 chr3      99638475   9.98e7 160752 +      ENSG00… protein_… COL8A1     1.90   4.26  222. 7.22e- 9
 6 chr6     168441151   1.69e8 232295 +      ENSG00… protein_… SMOC2      2.98   3.00  214. 8.72e- 9
 7 chr11    116835751   1.17e8   2200 -      ENSG00… protein_… APOA1     -1.91   6.89  202. 1.22e- 8
 8 chr20     36541497   3.66e7   9951 +      ENSG00… protein_… MYL9       2.21   9.09  193. 1.55e- 8
 9 chr10     76869601   7.76e7 768769 -      ENSG00… protein_… KCNMA1     2.91   4.04  191. 1.65e- 8
10 chr9      83938311   8.40e7  18676 -      ENSG00… protein_… C9orf64    5.73   1.84  176. 2.01e- 8
# ℹ 13,886 more rows
# ℹ 1 more variable: FDR &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="multiple-testing" class="slide level2">
<h2>Multiple Testing</h2>
<ul>
<li>Under H<sub>0</sub>: ~1 in 20 random p-values will be &lt; 0.05</li>
<li>With ~14,000 genes <span class="math inline">\(\implies\)</span> about 700 expected</li>
<li>These would be Type 1 errors <span class="math inline">\(\implies\)</span> false positives
<ul>
<li>We say something is happening, when it’s just noise</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>The False Discovery Rate (FDR) controls Type 1 errors at a given rate
<ul>
<li>Using <span class="math inline">\(\alpha = 0.05 \implies\)</span> happy with 5% of results being errors</li>
<li>Always just an estimate</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(res_M<span class="sc">$</span>FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1409</code></pre>
</div>
</div>
</div>
</section>
<section id="inspection-of-p-values" class="slide level2">
<h2>Inspection of P-values</h2>
<ul>
<li>Under H<sub>0</sub> <span class="math inline">\(\implies\)</span> <span class="math inline">\(p \sim \mathcal{U}(0, 1)\)</span></li>
<li>Under H<sub>A</sub> no distribution <span class="math inline">\(\implies\)</span> peak near zero</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_M <span class="sc">%&gt;%</span> </span>
<span id="cb59-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(PValue)) <span class="sc">+</span></span>
<span id="cb59-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.01</span>, <span class="at">fill =</span> <span class="st">"grey70"</span>, <span class="at">colour =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb59-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>))) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-pval-1.png">Click to show figure</a></p>
</div>
<aside class="notes">
<p>Looks as expected</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="volcano-plots" class="slide level2">
<h2>Volcano Plots</h2>
<ul>
<li>Plot developed during microarray analysis</li>
<li>Shows logFC against significance (<span class="math inline">\(p\)</span>-values)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_M <span class="sc">%&gt;%</span> </span>
<span id="cb60-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(logFC, <span class="sc">-</span><span class="fu">log10</span>(PValue), <span class="at">colour =</span> FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>)) <span class="sc">+</span></span>
<span id="cb60-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb60-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(</span>
<span id="cb60-5"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> gene_name),</span>
<span id="cb60-6"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> . <span class="sc">%&gt;%</span> </span>
<span id="cb60-7"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="fu">arrange</span>(PValue) <span class="sc">%&gt;%</span> </span>
<span id="cb60-8"><a href="" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb60-9"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb60-10"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb60-11"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-volcano-1.png">Click to show figure</a></p>
</div>
</section>
<section id="ma-plots" class="slide level2">
<h2>MA Plots</h2>
<ul>
<li>Two-colour microarray data introduced MA plots</li>
<li>Signal from two-colour arrays layered treat vs control in a single spot
<ul>
<li><code>A</code> for intensity averages</li>
<li><code>M</code> for mean intensity differences</li>
</ul></li>
<li>Still commonly used and the language remains</li>
<li>Sometimes called MD (mean-difference) plots
<ul>
<li><code>edgeR</code> provides <code>plotMD()</code></li>
</ul></li>
<li>Show average expression on the <code>x</code>-axis and logFC on <code>y</code>-axis
<ul>
<li>How highly expressed are any DE genes?</li>
</ul></li>
</ul>
</section>
<section id="ma-plots-1" class="slide level2">
<h2>MA Plots</h2>
<ul>
<li>Often start with an exploration then customise for our data</li>
</ul>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Initial Plot</a></li><li><a href="#tabset-2-2">Customised Plot</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_M <span class="sc">%&gt;%</span> </span>
<span id="cb61-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(logCPM, logFC)) <span class="sc">+</span></span>
<span id="cb61-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">colour =</span> FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>), <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb61-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-ma-1.png">Click to show figure</a></p>
</div>
</div>
<div id="tabset-2-2">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_M <span class="sc">%&gt;%</span> </span>
<span id="cb62-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(PValue)) <span class="sc">%&gt;%</span> <span class="co"># Why do this?</span></span>
<span id="cb62-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(logCPM, logFC, <span class="at">colour =</span> FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>)) <span class="sc">+</span></span>
<span id="cb62-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb62-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">colour =</span> <span class="st">'royalblue3'</span>, <span class="at">method =</span> <span class="st">"lm"</span>) <span class="sc">+</span></span>
<span id="cb62-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(</span>
<span id="cb62-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> gene_name),</span>
<span id="cb62-8"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> . <span class="sc">%&gt;%</span> </span>
<span id="cb62-9"><a href="" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">filter</span>(FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb62-10"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="fu">arrange</span>(<span class="fu">desc</span>(<span class="fu">abs</span>(logFC))) <span class="sc">%&gt;%</span> </span>
<span id="cb62-11"><a href="" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb62-12"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb62-13"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb62-14"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-ma-label-1.png">Click to show figure</a></p>
</div>
</div>
</div>
</div>
</section>
<section id="filtering-of-de-genes" class="slide level2">
<h2>Filtering of DE Genes</h2>
<ul>
<li>There is a tendency of the most extreme logFC to be in low expressed genes</li>
<li>The ratio of any number over a small number will <span class="math inline">\(\rightarrow \infty\)</span></li>
<li>If there are “too many” DE genes <span class="math inline">\(\implies\)</span> should we filter results?</li>
</ul>
<div class="fragment">
<ul>
<li>Early approaches were to filter on extreme logFC</li>
<li>Biased results towards low-expressed genes</li>
</ul>
</div>
</section>
<section id="range-based-hypothesis-testing" class="slide level2">
<h2>Range-Based Hypothesis Testing</h2>
<ul>
<li>Traditional H<sub>0</sub> is <span class="math inline">\(\mu = 0\)</span> against H<sub>A</sub> <span class="math inline">\(\mu \neq 0\)</span></li>
<li>An alternative is to use a range-based H<sub>0</sub> <span class="citation" data-cites="McCarthy2009-qf">(<a href="#/references" role="doc-biblioref" onclick="">McCarthy and Smyth 2009</a>)</span></li>
<li>Implemented in <code>glmTreat()</code></li>
<li>Default range is FC &gt; 20% in either direction</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_treatM <span class="ot">&lt;-</span> fit <span class="sc">%&gt;%</span> </span>
<span id="cb63-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glmTreat</span>(<span class="at">coef =</span> <span class="st">"treatmentM"</span>, <span class="at">lfc =</span> <span class="fu">log2</span>(<span class="fl">1.2</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb63-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">topTags</span>(<span class="at">n =</span> <span class="cn">Inf</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb63-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb63-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()</span>
<span id="cb63-6"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(res_treatM<span class="sc">$</span>FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 596</code></pre>
</div>
</div>
</section>
<section id="range-based-hypothesis-testing-1" class="slide level2">
<h2>Range-Based Hypothesis Testing</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_treatM <span class="sc">%&gt;%</span> </span>
<span id="cb65-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(PValue)) <span class="sc">%&gt;%</span> <span class="co"># Why do this?</span></span>
<span id="cb65-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(logCPM, logFC, <span class="at">colour =</span> FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>)) <span class="sc">+</span></span>
<span id="cb65-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb65-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">colour =</span> <span class="st">'royalblue3'</span>, <span class="at">method =</span> <span class="st">"lm"</span>) <span class="sc">+</span></span>
<span id="cb65-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(</span>
<span id="cb65-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label =</span> gene_name),</span>
<span id="cb65-8"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> . <span class="sc">%&gt;%</span> </span>
<span id="cb65-9"><a href="" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">filter</span>(FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb65-10"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="fu">arrange</span>(<span class="fu">desc</span>(<span class="fu">abs</span>(logFC))) <span class="sc">%&gt;%</span> </span>
<span id="cb65-11"><a href="" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb65-12"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span></span>
<span id="cb65-13"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb65-14"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/plot-ma-treat-1.png">Click to show figure</a></p>
</div>
<aside class="notes">
<p>Clearly some high logFC genes had weak statistical support</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="checking-individual-genes" class="slide level2">
<h2>Checking Individual Genes</h2>
<ul>
<li>Checking individual genes for changes in expression can be helpful</li>
<li>Choosing a small number helps with interpretability</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">cpm</span>(dge_filter, <span class="at">log =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb66-2"><a href="" aria-hidden="true" tabindex="-1"></a>  .[res_treatM<span class="sc">$</span>gene_id[<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>],] <span class="sc">%&gt;%</span> </span>
<span id="cb66-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"gene_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb66-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="fu">all_of</span>(<span class="st">"gene_id"</span>), <span class="at">names_to =</span> <span class="st">"id"</span>, <span class="at">values_to =</span> <span class="st">"logCPM"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb66-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(dge<span class="sc">$</span>genes) <span class="sc">%&gt;%</span> </span>
<span id="cb66-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(dge<span class="sc">$</span>samples) <span class="sc">%&gt;%</span></span>
<span id="cb66-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_name =</span> <span class="fu">fct_inorder</span>(gene_name)) <span class="sc">%&gt;%</span> </span>
<span id="cb66-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(treatment, logCPM, <span class="at">fill =</span> treatment)) <span class="sc">+</span></span>
<span id="cb66-9"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb66-10"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>gene_name, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb66-11"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/boxplot-topranked-1.png">Click to show figure</a></p>
</div>
</section>
<section id="heatmaps-across-multiple-genes" class="slide level2">
<h2>Heatmaps Across Multiple Genes</h2>
<ul>
<li>Showing larger groups of genes can also be informative</li>
<li>Enables very basic clustering of genes</li>
</ul>
<div class="fragment">
<ul>
<li>The package <code>pheatmap</code> is a common resource</li>
<li>Pre-dates the <code>tidyverse</code> <span class="math inline">\(\implies\)</span> <code>matrix</code> and <code>data.frame</code> objects</li>
</ul>
</div>
</section>
<section id="heatmaps-across-multiple-genes-1" class="slide level2">
<h2>Heatmaps Across Multiple Genes</h2>
<ol type="1">
<li>Form a matrix from the top genes</li>
<li>Subtract the mean expression value
<ul>
<li>This retains variability but shows direction of change</li>
</ul></li>
<li>Set rownames to be the gene-names</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge_filter, <span class="at">log =</span> <span class="cn">TRUE</span>)[res_treatM<span class="sc">$</span>gene_id[<span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>],] </span>
<span id="cb67-2"><a href="" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat <span class="sc">-</span> <span class="fu">rowMeans</span>(mat)</span>
<span id="cb67-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> <span class="fu">setNames</span>(genes<span class="sc">$</span>gene_name, genes<span class="sc">$</span>gene_id)[<span class="fu">rownames</span>(mat)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<ul>
<li>Choosing too many genes can confuse rather than illustrate the results!!!</li>
</ul>
</div>
</section>
<section id="heatmaps-across-multiple-genes-2" class="slide level2">
<h2>Heatmaps Across Multiple Genes</h2>
<ul>
<li>Groups can be set using column annotations (<code>annotation_col</code>)
<ul>
<li>Requires a <code>data.frame</code> with rownames</li>
</ul></li>
<li>Group colours can also be passed as a list</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(</span>
<span id="cb68-2"><a href="" aria-hidden="true" tabindex="-1"></a>  mat, </span>
<span id="cb68-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">annotation_col =</span> samples <span class="sc">%&gt;%</span></span>
<span id="cb68-4"><a href="" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(id, treatment) <span class="sc">%&gt;%</span> </span>
<span id="cb68-5"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb68-6"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">column_to_rownames</span>(<span class="st">"id"</span>),</span>
<span id="cb68-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">annotation_colors =</span> <span class="fu">list</span>(</span>
<span id="cb68-8"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">treatment =</span> <span class="fu">setNames</span>(</span>
<span id="cb68-9"><a href="" aria-hidden="true" tabindex="-1"></a>      RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">4</span>, <span class="st">"Set1"</span>), <span class="fu">levels</span>(samples<span class="sc">$</span>treatment)</span>
<span id="cb68-10"><a href="" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb68-11"><a href="" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb68-12"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">cutree_rows =</span> <span class="dv">4</span></span>
<span id="cb68-13"><a href="" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 75%;">
<p><a href="deg_slides_files/figure-revealjs/heatmap-topranked-1.png">Click to show figure</a></p>
</div>
<aside class="notes">
<p>Every column in the annotation df will be added as an annotation</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="using-deseq2" class="title-slide slide level1 center" data-background-color="#3c3c44" data-background-image="https://raw.githubusercontent.com/Bioconductor/BiocStickers/devel/DESeq2/DESeq2.png" data-background-size="30%" data-background-opacity="0.4" data-background-position="80% 50%">
<h1>Using DESeq2</h1>

</section>
<section id="deseq2" class="slide level2">
<h2>DESeq2</h2>
<ul>
<li>Another common package for RNASeq analysis is <code>DESeq2</code> <span class="citation" data-cites="DESeq22014">(<a href="#/references" role="doc-biblioref" onclick="">Love, Huber, and Anders 2014</a>)</span></li>
<li>Uses <code>S4</code> object more interconnected to other Bioconductor classes</li>
<li>Different approach to dispersions</li>
<li>Different statistical testing</li>
<li>Feels a bit more like a “<em>black-box</em>” to me
<ul>
<li>More default settings for a similar workflow <span class="math inline">\(\implies\)</span> fewer manual steps</li>
</ul></li>
<li>Is a very high-quality approach (as is <code>edgeR</code>)</li>
</ul>
</section>
<section id="deseq2-1" class="slide level2">
<h2>DESeq2</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb69-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(extraChIPs)</span>
<span id="cb69-3"><a href="" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb69-4"><a href="" aria-hidden="true" tabindex="-1"></a>  dge_filter<span class="sc">$</span>counts, </span>
<span id="cb69-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> samples,</span>
<span id="cb69-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowData =</span> dge_filter<span class="sc">$</span>genes <span class="sc">%&gt;%</span> </span>
<span id="cb69-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">makeGRangesFromDataFrame</span>(<span class="at">keep.extra.columns =</span> <span class="cn">TRUE</span>),</span>
<span id="cb69-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> treatment</span>
<span id="cb69-9"><a href="" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb69-10"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "DESeqDataSet"               "RangedSummarizedExperiment" "SummarizedExperiment"      
[4] "RectangularData"            "Vector"                     "Annotated"                 
[7] "vector_OR_Vector"          </code></pre>
</div>
</div>
</section>
<section id="summarized-experiment-objects" class="slide level2">
<h2>Summarized Experiment Objects</h2>
<ul>
<li><code>SummarizedExperiment</code> objects are a core Bioconductor class</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">getSlots</span>(<span class="st">"SummarizedExperiment"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            colData              assays               NAMES     elementMetadata            metadata 
        "DataFrame"    "Assays_OR_NULL" "character_OR_NULL"         "DataFrame"              "list" </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>Any number of <code>assays</code> can be added
<ul>
<li>Usually matrices with identical dimensions</li>
</ul></li>
<li><code>colData</code> holds the sample-level metadata</li>
</ul>
</div>
</section>
<section id="summarized-experiment-objects-1" class="slide level2">
<h2>Summarized Experiment Objects</h2>
<ul>
<li><code>RangedSummarizedExperiment</code> objects are a subclass of <code>SummarizedExperiment</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">getSlots</span>(<span class="st">"RangedSummarizedExperiment"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     rowRanges                        colData                         assays 
"GenomicRanges_OR_GRangesList"                    "DataFrame"               "Assays_OR_NULL" 
                         NAMES                elementMetadata                       metadata 
           "character_OR_NULL"                    "DataFrame"                         "list" </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>All slots from a <code>SummarizedExperiment</code></li>
<li><code>rowRanges</code> allow metadata for each row of counts
<ul>
<li>Usually a <code>GRanges</code> with optional <code>mcols</code></li>
</ul></li>
</ul>
</div>
</section>
<section id="deseqdataset-objects" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<ul>
<li>Extends <code>RangedSummarizedExperiment</code> objects</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">getSlots</span>(<span class="st">"DESeqDataSet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                        design             dispersionFunction                      rowRanges 
                         "ANY"                     "function" "GenomicRanges_OR_GRangesList" 
                       colData                         assays                          NAMES 
                   "DataFrame"               "Assays_OR_NULL"            "character_OR_NULL" 
               elementMetadata                       metadata 
                   "DataFrame"                         "list" </code></pre>
</div>
</div>
<div class="fragment">
<ul>
<li>Includes a design matrix &amp; dispersionFunction</li>
</ul>
</div>
</section>
<section id="deseqdataset-objects-1" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<ul>
<li>Multiple data access method exist
<ul>
<li><code>colData()</code>, <code>rowData()</code>, <code>rowRanges()</code></li>
<li><code>assay()</code>, <code>counts()</code></li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(dds)</span>
<span id="cb77-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">rowRanges</span>(dds)</span>
<span id="cb77-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(dds)</span>
<span id="cb77-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(dds, <span class="st">"counts"</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb77-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(dds) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="deseqdataset-objects-2" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<ul>
<li>All functions &amp; methods written for parent classes work with a <code>DESeqDataSet</code></li>
<li><code>tidyomics</code> specifically developed for <code>SummarizedExperiment</code> objects</li>
<li>Numerous functions &amp; methods for <code>SummarizedExperiment</code> objects</li>
</ul>
<div class="fragment">
<ul>
<li>Let’s look at a few from <code>extraChIPs</code></li>
<li>First add a <code>logCPM</code> assay</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(dds, <span class="st">"logCPM"</span>) <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dds, <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb78-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 2
names(2): counts logCPM</code></pre>
</div>
</div>
<aside class="notes">
<ul>
<li>I know the methods in <code>extraChIPs</code> best given I wrote them. Sorry…</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="deseqdataset-objects-3" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAssayPCA</span>(</span>
<span id="cb80-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dds, <span class="at">assay =</span> <span class="st">"logCPM"</span>, </span>
<span id="cb80-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">colour =</span> <span class="st">"treatment"</span>, <span class="at">label =</span> <span class="st">"id"</span></span>
<span id="cb80-4"><a href="" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb80-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">colour =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb80-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/ec-pca12-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="deseqdataset-objects-4" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAssayPCA</span>(</span>
<span id="cb81-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dds, <span class="at">assay =</span> <span class="st">"logCPM"</span>, </span>
<span id="cb81-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">colour =</span> <span class="st">"treatment"</span>, <span class="at">label =</span> <span class="st">"id"</span>,</span>
<span id="cb81-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">pc_x =</span> <span class="dv">2</span>, <span class="at">pc_y =</span> <span class="dv">3</span> <span class="co"># Change the PCs</span></span>
<span id="cb81-5"><a href="" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb81-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">colour =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb81-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/ec-pca23-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="deseqdataset-objects-5" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAssayDensities</span>(</span>
<span id="cb82-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dds, <span class="st">"logCPM"</span>, <span class="at">colour =</span> <span class="st">"treatment"</span></span>
<span id="cb82-3"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb82-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">colour =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb82-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/logcpm-densities-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="deseqdataset-objects-6" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAssayDensities</span>(</span>
<span id="cb83-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dds, <span class="at">colour =</span> <span class="st">"treatment"</span>, <span class="at">trans =</span> <span class="st">"log1p"</span></span>
<span id="cb83-3"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb83-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">colour =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb83-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/log1p-densities-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="deseqdataset-objects-7" class="slide level2">
<h2>DESeqDataSet Objects</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAssayRle</span>(</span>
<span id="cb84-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dds, <span class="st">"logCPM"</span>, </span>
<span id="cb84-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="st">"treatment"</span>, <span class="at">by_x =</span> <span class="st">"replicate"</span></span>
<span id="cb84-4"><a href="" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb84-5"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(</span>
<span id="cb84-6"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="sc">~</span>treatment, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">scales =</span> <span class="st">"free_x"</span></span>
<span id="cb84-7"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb84-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">"Treatment"</span>) <span class="sc">+</span></span>
<span id="cb84-9"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/logcpm-rle-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## RLE shows if any sample shows consistently</span></span>
<span id="cb85-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="do">## higher or lower counts</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<aside class="notes">
<ul>
<li>This makes another excellent QC check</li>
<li>Usually fine, but a good tool to know</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="dge-analysis-1" class="slide level2">
<h2>DGE Analysis</h2>
<ul>
<li>The function <code>DESeq()</code> wraps multiple steps</li>
<li>Normalisation factors <span class="math inline">\(\equiv\)</span> <code>sizeFactors</code> added to <code>colData</code>
<ul>
<li>Uses the RLE method (very similar to TMM)</li>
</ul></li>
<li>Additional assays will be added along with <code>rowData</code> columns</li>
<li>Testing is performed using the Wald Test</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb86-2"><a href="" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 13896 12 
metadata(1): version
assays(5): counts logCPM mu H cooks
rownames(13896): ENSG00000000003.14 ENSG00000000005.6 ... ENSG00000288022.1
  ENSG00000288066.1
rowData names(33): gene_id gene_type ... deviance maxCooks
colnames(12): 1_control 6_control ... 9_M 14_M
colData names(5): id replicate treatment group sizeFactor</code></pre>
</div>
</div>
</section>
<section id="dispersions" class="slide level2">
<h2>Dispersions</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="co"># Dispersions are also moderated by `DESeq2`</span></span>
<span id="cb88-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDispEsts</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/deseq-dispersions-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="dge-results" class="slide level2">
<h2>DGE Results</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Intercept"                 "treatment_M_vs_control"    "treatment_Nsp6_vs_control"
[4] "treatment_Nsp8_vs_control"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_deseq2_M <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">name =</span> <span class="st">"treatment_M_vs_control"</span>)</span>
<span id="cb91-2"><a href="" aria-hidden="true" tabindex="-1"></a>res_deseq2_M</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): treatment M vs control 
Wald test p-value: treatment M vs control 
DataFrame with 13896 rows and 6 columns
                    baseMean log2FoldChange     lfcSE       stat      pvalue        padj
                   &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003.14 4918.2962     -0.0734152 0.0548639  -1.338132 1.80853e-01 0.444613885
ENSG00000000005.6    80.3602     -0.8692770 0.3071633  -2.830016 4.65457e-03 0.036358343
ENSG00000000419.12 1710.4219      0.3642342 0.0843509   4.318081 1.57392e-05 0.000353798
ENSG00000000457.14  551.4245     -0.0386804 0.1648748  -0.234605 8.14516e-01 0.923290340
ENSG00000000460.17  190.9980      0.1536008 0.1744186   0.880645 3.78510e-01 0.650754388
...                      ...            ...       ...        ...         ...         ...
ENSG00000287966.1    43.9669     -0.5843052  0.448175 -1.3037444  0.19232074   0.4593943
ENSG00000287978.1    75.1925      0.0260297  0.412190  0.0631498  0.94964723   0.9831753
ENSG00000288012.1    32.8771     -0.2789995  0.462572 -0.6031481  0.54641019   0.7777428
ENSG00000288022.1    36.4753     -0.8772781  0.328395 -2.6714106  0.00755332   0.0523458
ENSG00000288066.1   438.5152      0.0771334  0.210274  0.3668230  0.71375102   0.8722577</code></pre>
</div>
</div>
</section>
<section id="dge-results-1" class="slide level2">
<h2>DGE Results</h2>
<ul>
<li>This analysis gave 1,972 DE genes</li>
<li>To sort by significance</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_deseq2_M <span class="sc">%&gt;%</span> </span>
<span id="cb93-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"gene_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb93-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="fu">as_tibble</span>(genes)) <span class="sc">%&gt;%</span> </span>
<span id="cb93-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(pvalue) <span class="sc">%&gt;%</span> </span>
<span id="cb93-5"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(range, <span class="fu">starts_with</span>(<span class="st">"gene"</span>), <span class="fu">everything</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 13,896 × 10
   range        gene_id gene_type gene_name baseMean log2FoldChange  lfcSE  stat    pvalue      padj
   &lt;chr&gt;        &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
 1 chr6:835528… ENSG00… protein_… SNAP91      2097.           -4.29 0.123  -34.9 2.49e-267 3.45e-263
 2 chr3:996384… ENSG00… protein_… COL8A1       727.            1.91 0.136   14.0 7.91e- 45 5.49e- 41
 3 chr6:168441… ENSG00… protein_… SMOC2        302.            2.98 0.215   13.9 1.21e- 43 5.61e- 40
 4 chr10:76869… ENSG00… protein_… KCNMA1       624.            2.91 0.212   13.8 3.59e- 43 1.25e- 39
 5 chr20:36541… ENSG00… protein_… MYL9       20813.            2.22 0.162   13.7 1.13e- 42 3.14e- 39
 6 chr6:846873… ENSG00… protein_… TBX18         97.4           4.65 0.340   13.7 1.41e- 42 3.26e- 39
 7 chr11:11683… ENSG00… protein_… APOA1       4526.           -1.90 0.143  -13.3 3.88e- 40 7.70e- 37
 8 chr9:164095… ENSG00… protein_… BNC2         257.            3.50 0.265   13.2 5.34e- 40 9.28e- 37
 9 chr22:17178… ENSG00… protein_… ADA2         734.           -1.20 0.0960 -12.5 7.89e- 36 1.22e- 32
10 chr6:560565… ENSG00… protein_… COL21A1      343.           -2.37 0.198  -12.0 5.38e- 33 7.48e- 30
# ℹ 13,886 more rows</code></pre>
</div>
</div>
<aside class="notes">
<p>This takes advantage of <code>as_tibble()</code> twice. Once to convert a <code>DataFrame</code>, with the second to convert a <code>GRanges</code></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="visualising-results" class="slide level2">
<h2>Visualising Results</h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMA</span>(res_deseq2_M)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="deg_slides_files/figure-revealjs/deseq-ma-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="filtering-results" class="slide level2">
<h2>Filtering Results</h2>
<div style="font-size: 90%;">
<ul>
<li>The approach taken by <code>DESeq2</code> is different from <code>edgeR</code>’s range-based H<sub>0</sub></li>
<li>Shrinks the estimates of logFC <span class="citation" data-cites="Zhu2019-kg">(<a href="#/references" role="doc-biblioref" onclick="">Zhu, Ibrahim, and Love 2019</a>)</span></li>
<li>Returns <code>s</code>-values (false sign or small)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_lfcshrink_M <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(</span>
<span id="cb96-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dds, <span class="st">"treatment_M_vs_control"</span>, <span class="at">lfcThreshold =</span> <span class="fu">log2</span>(<span class="fl">1.2</span>)</span>
<span id="cb96-3"><a href="" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb96-4"><a href="" aria-hidden="true" tabindex="-1"></a>res_lfcshrink_M</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MAP): treatment M vs control 
 
DataFrame with 13896 rows and 4 columns
                    baseMean log2FoldChange     lfcSE    svalue
                   &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSG00000000003.14 4918.2962     -0.0675130 0.0529894 0.7792084
ENSG00000000005.6    80.3602     -0.6065355 0.3664508 0.0413765
ENSG00000000419.12 1710.4219      0.3378204 0.0851769 0.0458250
ENSG00000000457.14  551.4245     -0.0192876 0.1236106 0.6995773
ENSG00000000460.17  190.9980      0.0854692 0.1358403 0.5657148
...                      ...            ...       ...       ...
ENSG00000287966.1    43.9669    -0.09359097  0.200388 0.4204071
ENSG00000287978.1    75.1925     0.00454662  0.168958 0.6258674
ENSG00000288012.1    32.8771    -0.03896311  0.176866 0.5545518
ENSG00000288022.1    36.4753    -0.56074752  0.410878 0.0593301
ENSG00000288066.1   438.5152     0.03309121  0.140818 0.6493741</code></pre>
</div>
</div>
</div>
</section>
<section id="filtering-results-1" class="slide level2">
<h2>Filtering Results</h2>
<ul>
<li>This has reduced the original 1,972 DE genes to 1,270</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_lfcshrink_M <span class="sc">%&gt;%</span> </span>
<span id="cb98-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"gene_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb98-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="fu">as_tibble</span>(genes)) <span class="sc">%&gt;%</span> </span>
<span id="cb98-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(svalue) <span class="sc">%&gt;%</span> </span>
<span id="cb98-5"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(svalue <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb98-6"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(range, <span class="fu">starts_with</span>(<span class="st">"gene"</span>), <span class="fu">everything</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,270 × 8
   range                       gene_id   gene_type gene_name baseMean log2FoldChange  lfcSE   svalue
   &lt;chr&gt;                       &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;
 1 chr6:83552880-83709691:-    ENSG0000… protein_… SNAP91      2097.          -4.28  0.123  0       
 2 chr22:17178790-17258235:-   ENSG0000… protein_… ADA2         734.          -1.18  0.0966 0       
 3 chr11:116835751-116837950:- ENSG0000… protein_… APOA1       4526.          -1.88  0.144  0       
 4 chr6:56056590-56394094:-    ENSG0000… protein_… COL21A1      343.          -2.33  0.199  0       
 5 chr8:33591330-33600023:-    ENSG0000… protein_… DUSP26        73.3         -6.34  0.597  0       
 6 chr9:128255829-128275995:-  ENSG0000… protein_… GOLGA2      1615.          -0.942 0.0817 0       
 7 chr6:84687351-84764598:-    ENSG0000… protein_… TBX18         97.4          4.60  0.338  9.79e-39
 8 chr6:168441151-168673445:+  ENSG0000… protein_… SMOC2        302.           2.95  0.216  1.28e-36
 9 chr10:76869601-77638369:-   ENSG0000… protein_… KCNMA1       624.           2.88  0.212  5.28e-36
10 chr9:16409503-16870843:-    ENSG0000… protein_… BNC2         257.           3.46  0.266  1.29e-34
# ℹ 1,260 more rows</code></pre>
</div>
</div>
</section>
<section id="comparison-of-approaches" class="slide level2">
<h2>Comparison of Approaches</h2>
<ul>
<li>Using the most conservative analyses</li>
<li>Genes along zero for <code>DESeq2</code> were a result of shrinkage</li>
</ul>

<img data-src="deg_slides_files/figure-revealjs/plot-compare-1.png" class="quarto-figure quarto-figure-left r-stretch" width="768"></section>
<section id="export-results" class="slide level2">
<h2>Export Results</h2>
<ul>
<li>Save the <code>edgeR</code> results using the following</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">write_rds</span>(res_treatM, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data/res_treatM.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="conclusion" class="slide level2">
<h2>Conclusion</h2>
<ul>
<li><code>DESeq2</code> data structure integrate nicely with other Bioconductor packages</li>
<li><code>edgeR</code> can also handle <code>SummarizedExperiment</code> objects (mostly)</li>
<li>No right or wrong method or data classes to use
<ul>
<li>Results were still broadly consistent</li>
</ul></li>
<li><code>DESeq2</code> also has many options for parameter/model choice not covered here</li>
</ul>
<div class="fragment">
<ul>
<li><code>SummarizedExperiment</code> objects share much with <code>SingleCellExperiment</code> objects</li>
<li>Have helped with multiple analyses &amp; seem preferable to <code>Seurat</code> (IMHO)
<ul>
<li><code>Seurat</code> is everywhere but I spend a lot of time debugging for people</li>
</ul></li>
</ul>
</div>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<div class="quarto-auto-generated-content">
<p><img src="assets/bodl_logo_white_background.jpg" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Liao2014-gy" class="csl-entry" role="listitem">
Liao, Yang, Gordon K Smyth, and Wei Shi. 2014. <span>“featureCounts: An Efficient General Purpose Program for Assigning Sequence Reads to Genomic Features.”</span> <em>Bioinformatics</em> 30 (7): 923–30.
</div>
<div id="ref-Liu2023-qu" class="csl-entry" role="listitem">
Liu, Juli, Shiyong Wu, Yucheng Zhang, Cheng Wang, Sheng Liu, Jun Wan, and Lei Yang. 2023. <span>“<span>SARS-CoV-2</span> Viral Genes Nsp6, Nsp8, and <span>M</span> Compromise Cellular <span>ATP</span> Levels to Impair Survival and Function of Human Pluripotent Stem Cell-Derived Cardiomyocytes.”</span> <em>Stem Cell Res. Ther.</em> 14 (1): 249.
</div>
<div id="ref-DESeq22014" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Lund2012-xo" class="csl-entry" role="listitem">
Lund, Steven P, Dan Nettleton, Davis J McCarthy, and Gordon K Smyth. 2012. <span>“Detecting Differential Expression in <span class="nocase">RNA-sequence</span> Data Using Quasi-Likelihood with Shrunken Dispersion Estimates.”</span> <em>Stat. Appl. Genet. Mol. Biol.</em> 11 (5).
</div>
<div id="ref-McCarthy2009-qf" class="csl-entry" role="listitem">
McCarthy, Davis J, and Gordon K Smyth. 2009. <span>“Testing Significance Relative to a Fold-Change Threshold Is a <span>TREAT</span>.”</span> <em>Bioinformatics</em> 25 (6): 765–71.
</div>
<div id="ref-edgeR2010" class="csl-entry" role="listitem">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. <span>“edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-Robinson2010-qp" class="csl-entry" role="listitem">
Robinson, Mark D, and Alicia Oshlack. 2010. <span>“A Scaling Normalization Method for Differential Expression Analysis of <span class="nocase">RNA-seq</span> Data.”</span> <em>Genome Biol.</em> 11 (3): R25.
</div>
<div id="ref-Zhu2019-kg" class="csl-entry" role="listitem">
Zhu, Anqi, Joseph G Ibrahim, and Michael I Love. 2019. <span>“Heavy-Tailed Prior Distributions for Sequence Count Data: Removing the Noise and Preserving Large Differences.”</span> <em>Bioinformatics</em> 35 (12): 2084–92.
</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="deg_slides_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="deg_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="deg_slides_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="deg_slides_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="deg_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="deg_slides_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="deg_slides_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="deg_slides_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="deg_slides_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="deg_slides_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    <script type="text/javascript">
      Reveal.on('ready', event => {
        if (event.indexh === 0) {
          document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
        }
      });
      Reveal.addEventListener('slidechanged', (event) => {
        if (event.indexh === 0) {
          Reveal.configure({ slideNumber: null });
          document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
        }
        if (event.indexh === 1) {
          Reveal.configure({ slideNumber: 'c' });
          document.querySelector("div.has-logo > img.slide-logo").style.display = null;
        }
      });
    </script>
    

</body></html>