<!DOCTYPE html>
<html lang="en"><head>
<script src="enrichment_slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="enrichment_slides_files/libs/quarto-html/tabby.min.js"></script>
<script src="enrichment_slides_files/libs/quarto-html/popper.min.js"></script>
<script src="enrichment_slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="enrichment_slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="enrichment_slides_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="enrichment_slides_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="enrichment_slides_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.4.555">

  <meta name="author" content="Dr Stevie Pederson">
  <meta name="dcterms.date" content="2024-07-11">
  <title>Finding the Underlying Biology</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="enrichment_slides_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="enrichment_slides_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="enrichment_slides_files/libs/revealjs/dist/theme/quarto.css">
  <link href="enrichment_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="enrichment_slides_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="enrichment_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="enrichment_slides_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-color="#3d3d40" data-background-image="assets/bodl_logo_white_background.jpg" data-background-opacity="0.3" data-background-size="90%" class="quarto-title-block center">
  <h1 class="title">Finding the Underlying Biology</h1>
  <p class="subtitle">RAdelaide 2024</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Dr Stevie Pederson 
</div>
        <p class="quarto-title-affiliation">
            </p><div class="line-block">Black Ochre Data Labs<br>
            Telethon Kids Institute</div>
          <p></p>
    </div>
</div>

  <p class="date">July 11, 2024</p>
</section>
<section>
<section id="enrichment-analysis" class="title-slide slide level1 center" data-background-color="#3c3c44">
<h1>Enrichment Analysis</h1>

</section>
<section id="what-is-enrichment-analysis" class="slide level2">
<h2>What is Enrichment Analysis</h2>
<ul>
<li>Often in transcriptomics and genomics we have large lists of results</li>
<li>May be genes, genomic positions etc</li>
<li>How do we interpret these?
<ul>
<li>Should we cherry pick known <em>high-quality</em> candidates?</li>
<li>Should we try look for patterns within the list?</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Enrichment analysis is one strategy for looking for patterns
<ul>
<li>Network analysis is another approach</li>
</ul></li>
<li>Generally rely on genes can be organised into “gene sets”
<ul>
<li>Multiple Databases depending on your question</li>
</ul></li>
</ul>
</div>
</section>
<section id="multiple-approaches" class="slide level2">
<h2>Multiple Approaches</h2>
<ul>
<li>Take a ranked list of results<br><span class="math inline">\(\implies\)</span>Are gene-sets clustered around either end</li>
</ul>
<div class="fragment">
<ul>
<li>Are genes from a specific gene-set enriched within significant results
<ul>
<li>Hypergeometric testing</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Can we obtain an overall value of significance for all genes within a gene-set</li>
</ul>
<aside class="notes">
<ul>
<li>No method is perfect &amp; all have limitations</li>
<li>Gene expression also appears to be highly correlated &amp; subject to bias</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="gene-ontology" class="slide level2">
<h2>Gene Ontology</h2>
<ul>
<li>Gene Ontology (GO) Database https://www.geneontology.org/</li>
<li>Restricted and highly controlled set of descriptive terms</li>
<li>Terms are grouped into 3 main categories
<ol type="1">
<li>Biological Process (BP)</li>
<li>Molecular Function (MF)</li>
<li>Cellular Component (CC)</li>
</ol></li>
<li>Hierarchical interconnected structure within each category</li>
</ul>
</section>
<section id="gene-ontology-1" class="slide level2">
<h2>Gene Ontology</h2>

<img data-src="https://www.geneontology.org/assets/ontology-browse-tree.jpg" class="r-stretch quarto-figure-center"><p class="caption">Taken from https://www.geneontology.org</p><aside class="notes">
<ul>
<li>Knowing how to filter results with heavily related terms can be a challenge</li>
<li>Multiple pruning strategies exist</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="gene-ontology-2" class="slide level2">
<h2>Gene Ontology</h2>
<ul>
<li>An example using the QuickGO tool is <a href="https://www.ebi.ac.uk/QuickGO/GTerm?id=GO:0002253">here</a></li>
</ul>
</section>
<section id="kegg" class="slide level2">
<h2>KEGG</h2>
<ul>
<li>Kyoto Encyclopedia of Genes and Genomes (KEGG)</li>
<li>Exhaustive list of manually curated pathways
<ol type="1">
<li>Metabolism</li>
<li>Genetic Information Processing</li>
<li>Environmental Information Processing</li>
<li>Cellular Processes</li>
<li>Organismal Systems</li>
<li>Human Diseases</li>
<li>Drug Development</li>
</ol></li>
<li>Role in pathway (e.g.&nbsp;activation of inhibition) included</li>
</ul>
</section>
<section id="kegg-1" class="slide level2">
<h2>KEGG</h2>

<img data-src="assets/kegg_mapk.png" class="r-stretch quarto-figure-left"><p class="caption">https://www.genome.jp/pathway/hsa04010</p></section>
<section id="other-databases" class="slide level2">
<h2>Other Databases</h2>
<ul>
<li>WikiPathways is an open-source analog to KEGG
<ul>
<li>https://www.wikipathways.org</li>
</ul></li>
<li>Reactome is highly curated, peer-reviewed database
<ul>
<li>https://reactome.org</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Can also use custom gene-sets
<ul>
<li>Results from an internal study</li>
<li>Genes matching custom genomic features</li>
</ul></li>
</ul>
</div>
</section>
<section id="msigdb" class="slide level2">
<h2>MSigDB</h2>
<ul>
<li>The Molecular Signatures Database (MSigDB) collects multiple databases
<ul>
<li>https://www.gsea-msigdb.org/gsea/msigdb/index.jsp</li>
<li>Collected simply as gene sets without functional annotation</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>H: Hallmark Gene Sets are unique to MSigDB</li>
</ul>
</div>
</section>
<section id="gene-sets-and-pathways-in-r" class="slide level2">
<h2>Gene Sets and Pathways in R</h2>
<ul>
<li>Bioconductor provides multiple packages and tools for obtaining gene sets
<ul>
<li><code>GO.db</code>, <code>KEGGREST</code>, <code>biomaRt</code></li>
</ul></li>
<li>My preferred strategy is <code>msigdbr</code>
<ul>
<li>Parses MSigDb gene sets as a <code>tibble</code></li>
<li>Not really built on existing Bioconductor object classes</li>
</ul></li>
<li>Best annotations are for human/mouse</li>
</ul>
</section>
<section id="msigdbr" class="slide level2">
<h2><code>msigdbr</code></h2>
<div class="cell columns column-output-location" data-layout-align="center">
<div class="column">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span>
<span id="cb1-3"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(msigdbr)</span>
<span id="cb1-4"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">msigdbr_collections</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 23 × 3
   gs_cat gs_subcat         num_genesets
   &lt;chr&gt;  &lt;chr&gt;                    &lt;int&gt;
 1 C1     ""                         299
 2 C2     "CGP"                     3384
 3 C2     "CP"                        29
 4 C2     "CP:BIOCARTA"              292
 5 C2     "CP:KEGG"                  186
 6 C2     "CP:PID"                   196
 7 C2     "CP:REACTOME"             1615
 8 C2     "CP:WIKIPATHWAYS"          664
 9 C3     "MIR:MIRDB"               2377
10 C3     "MIR:MIR_Legacy"           221
# ℹ 13 more rows</code></pre>
</div>
</div>
</div>
</section>
<section id="msigdbr-1" class="slide level2">
<h2><code>msigdbr</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">msigdbr_species</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 2
   species_name                    species_common_name                          
   &lt;chr&gt;                           &lt;chr&gt;                                        
 1 Anolis carolinensis             Carolina anole, green anole                  
 2 Bos taurus                      bovine, cattle, cow, dairy cow, domestic cat…
 3 Caenorhabditis elegans          &lt;NA&gt;                                         
 4 Canis lupus familiaris          dog, dogs                                    
 5 Danio rerio                     leopard danio, zebra danio, zebra fish, zebr…
 6 Drosophila melanogaster         fruit fly                                    
 7 Equus caballus                  domestic horse, equine, horse                
 8 Felis catus                     cat, cats, domestic cat                      
 9 Gallus gallus                   bantam, chicken, chickens, Gallus domesticus 
10 Homo sapiens                    human                                        
11 Macaca mulatta                  rhesus macaque, rhesus macaques, Rhesus monk…
12 Monodelphis domestica           gray short-tailed opossum                    
13 Mus musculus                    house mouse, mouse                           
14 Ornithorhynchus anatinus        duck-billed platypus, duckbill platypus, pla…
15 Pan troglodytes                 chimpanzee                                   
16 Rattus norvegicus               brown rat, Norway rat, rat, rats             
17 Saccharomyces cerevisiae        baker's yeast, brewer's yeast, S. cerevisiae 
18 Schizosaccharomyces pombe 972h- &lt;NA&gt;                                         
19 Sus scrofa                      pig, pigs, swine, wild boar                  
20 Xenopus tropicalis              tropical clawed frog, western clawed frog    </code></pre>
</div>
</div>
</section>
<section id="setting-up-an-object" class="slide level2">
<h2>Setting Up An Object</h2>
<ul>
<li>Let’s choose KEGG pathways for our analysis</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="" aria-hidden="true" tabindex="-1"></a>kegg_tbl <span class="ot">&lt;-</span> <span class="fu">msigdbr</span>(<span class="at">subcategory =</span> <span class="st">"CP:KEGG"</span>)</span>
<span id="cb5-2"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(kegg_tbl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 16,283
Columns: 15
$ gs_cat             &lt;chr&gt; "C2", "C2", "C2", "C2", "C2", "C2", "C2", "C2", "C2…
$ gs_subcat          &lt;chr&gt; "CP:KEGG", "CP:KEGG", "CP:KEGG", "CP:KEGG", "CP:KEG…
$ gs_name            &lt;chr&gt; "KEGG_ABC_TRANSPORTERS", "KEGG_ABC_TRANSPORTERS", "…
$ gene_symbol        &lt;chr&gt; "ABCA1", "ABCA10", "ABCA12", "ABCA13", "ABCA2", "AB…
$ entrez_gene        &lt;int&gt; 19, 10349, 26154, 154664, 20, 21, 24, 23461, 23460,…
$ ensembl_gene       &lt;chr&gt; "ENSG00000165029", "ENSG00000154263", "ENSG00000144…
$ human_gene_symbol  &lt;chr&gt; "ABCA1", "ABCA10", "ABCA12", "ABCA13", "ABCA2", "AB…
$ human_entrez_gene  &lt;int&gt; 19, 10349, 26154, 154664, 20, 21, 24, 23461, 23460,…
$ human_ensembl_gene &lt;chr&gt; "ENSG00000165029", "ENSG00000154263", "ENSG00000144…
$ gs_id              &lt;chr&gt; "M11911", "M11911", "M11911", "M11911", "M11911", "…
$ gs_pmid            &lt;chr&gt; "", "", "", "", "", "", "", "", "", "", "", "", "",…
$ gs_geoid           &lt;chr&gt; "", "", "", "", "", "", "", "", "", "", "", "", "",…
$ gs_exact_source    &lt;chr&gt; "hsa02010", "hsa02010", "hsa02010", "hsa02010", "hs…
$ gs_url             &lt;chr&gt; "http://www.genome.jp/kegg/pathway/hsa/hsa02010.htm…
$ gs_description     &lt;chr&gt; "ABC transporters", "ABC transporters", "ABC transp…</code></pre>
</div>
</div>
</section>
<section id="setting-up-an-object-1" class="slide level2">
<h2>Setting Up An Object</h2>
<ul>
<li>Every mapping between a gene ID and KEGG pathway is here</li>
<li>Gene IDs are provided using Ensembl (EBI) and EntrezGene (NCBI)</li>
<li>Gene set IDs match the specific database</li>
<li>URLs are also provided
<ul>
<li>But can get out of date when databases reconfigure</li>
</ul></li>
</ul>
</section>
<section id="comparing-to-our-data" class="slide level2">
<h2>Comparing To Our Data</h2>
<div style="font-size: 95%;">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="" aria-hidden="true" tabindex="-1"></a>res_treatM <span class="ot">&lt;-</span> <span class="fu">read_rds</span>(<span class="st">"data/res_treatM.rds"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb7-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ensembl_gene =</span> <span class="fu">str_extract</span>(gene_id, <span class="st">"ENSG[0-9]+"</span>)) <span class="co"># Remove version numbers</span></span>
<span id="cb7-3"><a href="" aria-hidden="true" tabindex="-1"></a>kegg_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb7-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb7-5"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">gs_size =</span> dplyr<span class="sc">::</span><span class="fu">n</span>(),</span>
<span id="cb7-6"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">detected =</span> <span class="fu">sum</span>(ensembl_gene <span class="sc">%in%</span> res_treatM<span class="sc">$</span>ensembl_gene),</span>
<span id="cb7-7"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> gs_name</span>
<span id="cb7-8"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb7-9"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prop_detected =</span> detected <span class="sc">/</span> gs_size) <span class="sc">%&gt;%</span> </span>
<span id="cb7-10"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(prop_detected)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 186 × 4
   gs_name                                        gs_size detected prop_detected
   &lt;chr&gt;                                            &lt;int&gt;    &lt;int&gt;         &lt;dbl&gt;
 1 KEGG_ASTHMA                                        113        2        0.0177
 2 KEGG_GRAFT_VERSUS_HOST_DISEASE                     314        7        0.0223
 3 KEGG_OLFACTORY_TRANSDUCTION                        523       19        0.0363
 4 KEGG_AUTOIMMUNE_THYROID_DISEASE                    166        7        0.0422
 5 KEGG_ALLOGRAFT_REJECTION                           157        7        0.0446
 6 KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION           535       30        0.0561
 7 KEGG_TYPE_I_DIABETES_MELLITUS                      170       12        0.0706
 8 KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUC…     125        9        0.072 
 9 KEGG_TASTE_TRANSDUCTION                             89       10        0.112 
10 KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY     465       62        0.133 
# ℹ 176 more rows</code></pre>
</div>
</div>
</div>
</section>
<section id="comparing-to-our-data-1" class="slide level2">
<h2>Comparing To Our Data</h2>
<ul>
<li>Gene sets should be chosen wisely
<ul>
<li>If few genes from a gene set are detected<br><span class="math inline">\(\implies\)</span>may not be suitable for cell-type under investigation</li>
</ul></li>
<li>Should we restrict gene-sets to genes detected in our data?</li>
</ul>
<div class="fragment">
<ul>
<li>This can impact the analysis depending on the approach taken</li>
</ul>
<aside class="notes">
<ul>
<li>My take is that we should. For enrichment testing then we’re comparing changed vs unchanged in the candidate genes. Otherwise, we’re confounding detected with changed</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
<section id="cleaning-the-database" class="slide level2">
<h2>Cleaning The Database</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="" aria-hidden="true" tabindex="-1"></a>kegg_tbl <span class="ot">&lt;-</span> kegg_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb9-2"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb9-3"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sum</span>(ensembl_gene <span class="sc">%in%</span> res_treatM<span class="sc">$</span>ensembl_gene) <span class="sc">&gt;</span>  <span class="fl">0.05</span> <span class="sc">*</span> dplyr<span class="sc">::</span><span class="fu">n</span>(),</span>
<span id="cb9-4"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> gs_name</span>
<span id="cb9-5"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb9-6"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb9-7"><a href="" aria-hidden="true" tabindex="-1"></a>    ensembl_gene <span class="sc">%in%</span> res_treatM<span class="sc">$</span>ensembl_gene</span>
<span id="cb9-8"><a href="" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-cautionary-limitation" class="slide level2">
<h2>A Cautionary Limitation</h2>
<ul>
<li>Not all genes are annotated to a gene set or pathway</li>
<li>Well-studied genes are annotated to gene sets <br><span class="math inline">\(\implies\)</span> their pathways become interesting <br><span class="math inline">\(\implies\)</span> studied further<br><span class="math inline">\(\implies\)</span> rinse &amp; repeat</li>
</ul>
<div class="fragment">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(kegg_tbl, gene_symbol <span class="sc">==</span> <span class="st">"SNAP91"</span>) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(kegg_tbl, gene_symbol <span class="sc">==</span> <span class="st">"DUSP26"</span>) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
</div>
<div class="fragment">
<ul>
<li>This doesn’t mean these genes play an unimportant biological role!</li>
</ul>
<aside class="notes">
<ul>
<li>We can only use the available information</li>
<li>Poorly studied genes can be revelations but take large sums of money</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section></section>
<section>
<section id="gsea" class="title-slide slide level1 center" data-background-color="#3c3c44">
<h1>GSEA</h1>

</section>
<section id="gene-set-enrichment-analysis" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<ul>
<li>GSEA is a formally defined algorithm <span class="citation" data-cites="Subramanian2005-lx">(<a href="#/references" role="doc-biblioref" onclick="">Subramanian et al. 2005</a>)</span>
<ul>
<li>Not commonly used in reference to the wider type of analysis</li>
</ul></li>
<li>Doesn’t require DEGs <span class="math inline">\(\implies\)</span> uses a ranked list
<ul>
<li>Ranks based on significance only or incorporating direction</li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Walks along the gene list obtaining a cumulative sum
<ul>
<li>Can be set to start from either end</li>
<li>Genes in a gene-set return a +ve score</li>
<li>Genes not in a gene-set return a -ve score</li>
</ul></li>
<li>Compared to permuted values to return <span class="math inline">\(p\)</span>-values
<ul>
<li>Results can change on repeat runs</li>
</ul></li>
</ul>
</div>
</section>
<section id="gene-set-enrichment-analysis-1" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<ul>
<li>Can be run as a stand-alone tools</li>
<li><code>R</code> implementation is <code>fgsea</code> <span class="citation" data-cites="fgsea2019">(<a href="#/references" role="doc-biblioref" onclick="">Korotkevich, Sukhov, and Sergushichev 2019</a>)</span></li>
<li>Requires two inputs
<ul>
<li>A <code>list</code> with genes separated by gene set</li>
<li>A ranked-list of genes</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fgsea)</span>
<span id="cb14-2"><a href="" aria-hidden="true" tabindex="-1"></a>kegg_by_gs <span class="ot">&lt;-</span> kegg_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb14-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">split</span>(.<span class="sc">$</span>gs_name) <span class="sc">%&gt;%</span> </span>
<span id="cb14-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(pull, <span class="st">"ensembl_gene"</span>)</span>
<span id="cb14-5"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(kegg_by_gs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 181</code></pre>
</div>
</div>
</section>
<section id="gene-set-enrichment-analysis-2" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<ul>
<li>To make a directional ranked list
<ul>
<li>Multiply <span class="math inline">\(-\log_{10}p\)</span> by the sign of logFC</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="" aria-hidden="true" tabindex="-1"></a>ranks_treatM <span class="ot">&lt;-</span> res_treatM <span class="sc">%&gt;%</span> </span>
<span id="cb16-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb16-3"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">rank =</span> <span class="sc">-</span><span class="fu">sign</span>(logFC) <span class="sc">*</span> <span class="fu">log10</span>(PValue),</span>
<span id="cb16-4"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">rank =</span> <span class="fu">setNames</span>(rank, ensembl_gene)</span>
<span id="cb16-5"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb16-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(rank) <span class="sc">%&gt;%</span> </span>
<span id="cb16-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(rank)</span>
<span id="cb16-8"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ranks_treatM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ENSG00000065609 ENSG00000133878 ENSG00000118137 ENSG00000124749 ENSG00000093072 
     -12.316058      -10.005216       -7.726357       -7.383672       -7.189801 
ENSG00000167110 
      -6.815165 </code></pre>
</div>
</div>
</section>
<section id="gene-set-enrichment-analysis-3" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">101</span>)</span>
<span id="cb18-2"><a href="" aria-hidden="true" tabindex="-1"></a>gsea_treatM <span class="ot">&lt;-</span> <span class="fu">fgsea</span>(kegg_by_gs, ranks_treatM)</span>
<span id="cb18-3"><a href="" aria-hidden="true" tabindex="-1"></a>gsea_treatM <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(pval)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                     pathway         pval         padj
                                      &lt;char&gt;        &lt;num&gt;        &lt;num&gt;
  1:                           KEGG_LYSOSOME 1.224437e-07 2.216231e-05
  2:           KEGG_ECM_RECEPTOR_INTERACTION 9.315488e-05 6.322486e-03
  3:                     KEGG_FOCAL_ADHESION 1.047926e-04 6.322486e-03
  4:    KEGG_DRUG_METABOLISM_CYTOCHROME_P450 9.649298e-04 4.366307e-02
  5: KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION 1.622066e-03 5.871881e-02
 ---                                                                  
177:            KEGG_CITRATE_CYCLE_TCA_CYCLE 1.000000e+00 1.000000e+00
178:                KEGG_HUNTINGTONS_DISEASE 1.000000e+00 1.000000e+00
179:                           KEGG_RIBOSOME 1.000000e+00 1.000000e+00
180:                    KEGG_RNA_DEGRADATION 1.000000e+00 1.000000e+00
181:                        KEGG_SPLICEOSOME 1.000000e+00 1.000000e+00
        log2err         ES        NES  size
          &lt;num&gt;      &lt;num&gt;      &lt;num&gt; &lt;int&gt;
  1: 0.69013246  0.7163092  1.8838119   103
  2: 0.53843410  0.7019563  1.7491321    65
  3: 0.53843410  0.5897077  1.6219153   162
  4: 0.47727082 -0.7902400 -1.7551751    21
  5: 0.45505987  0.6242822  1.5895677    77
 ---                                       
177: 0.03607300  0.2037979  0.4382494    27
178: 0.08730983 -0.2367126 -0.7463800   160
179: 0.02921031  0.1473646  0.3766860    79
180: 0.06538342 -0.1415776 -0.3838046    52
181: 0.02555230  0.1980841  0.5294166   121
                                                                                             leadingEdge
                                                                                                  &lt;list&gt;
  1: ENSG00000103811,ENSG00000064601,ENSG00000152056,ENSG00000107798,ENSG00000078081,ENSG00000179163,...
  2: ENSG00000186340,ENSG00000053747,ENSG00000041982,ENSG00000081052,ENSG00000132470,ENSG00000060718,...
  3: ENSG00000101335,ENSG00000186340,ENSG00000053747,ENSG00000041982,ENSG00000081052,ENSG00000166501,...
  4: ENSG00000130649,ENSG00000069535,ENSG00000134184,ENSG00000085871,ENSG00000143198,ENSG00000076258,...
  5: ENSG00000156113,ENSG00000101335,ENSG00000096433,ENSG00000166501,ENSG00000163017,ENSG00000065675,...
 ---                                                                                                    
177:                     ENSG00000138413,ENSG00000131473,ENSG00000122729,ENSG00000173599,ENSG00000166411
178: ENSG00000146592,ENSG00000198959,ENSG00000137841,ENSG00000285437,ENSG00000124172,ENSG00000228049,...
179: ENSG00000171858,ENSG00000105372,ENSG00000140988,ENSG00000185088,ENSG00000137154,ENSG00000149806,...
180: ENSG00000123737,ENSG00000272886,ENSG00000166889,ENSG00000108515,ENSG00000114127,ENSG00000138767,...
181:                                     ENSG00000132792,ENSG00000173110,ENSG00000101161,ENSG00000111786</code></pre>
</div>
</div>
<aside class="notes">
<ul>
<li>Using <code>set.seed()</code> can aid reproducibility as the same random seed is used for permutations</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="gene-set-enrichment-analysis-4" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<ul>
<li>Adjusted p-values are FDR-adjusted by default</li>
<li>Enrichment Score (ES) is the peak value from the cumulative sum
<ul>
<li>ES &lt; 0 <span class="math inline">\(\implies\)</span> result is for genes at the negative end</li>
</ul></li>
<li>Normalised Enrichment Score (NES) normalises across pathways</li>
<li>The size of the gene set is also given</li>
<li>The Leading Edge lists the genes more beyond the peak ES
<ul>
<li>Might need to convert to gene names</li>
</ul></li>
</ul>
</section>
<section id="gene-set-enrichment-analysis-5" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<ul>
<li>The enrichment score can be seen for any pathway</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">plotEnrichment</span>(kegg_by_gs<span class="sc">$</span>KEGG_LYSOSOME, ranks_treatM) <span class="sc">+</span></span>
<span id="cb20-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"KEGG_LYSOSOME"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="enrichment_slides_files/figure-revealjs/lysosome-barcode-1.png" class="quarto-figure quarto-figure-center r-stretch" width="768"></section>
<section id="gene-set-enrichment-analysis-6" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<ul>
<li>Manual inspection can help us decide if we really believe the result
<ul>
<li>GSEA tends <em>always</em> find something <span class="math inline">\(\implies\)</span> be conservative with results</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb21-2"><a href="" aria-hidden="true" tabindex="-1"></a>gsea_treatM <span class="sc">%&gt;%</span> </span>
<span id="cb21-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(pval) <span class="sc">%&gt;%</span> </span>
<span id="cb21-4"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb21-5"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>) <span class="sc">%&gt;%</span> <span class="co"># Ensure we don't get more than 9</span></span>
<span id="cb21-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(pathway) <span class="sc">%&gt;%</span> </span>
<span id="cb21-7"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb21-8"><a href="" aria-hidden="true" tabindex="-1"></a>    \(x) {</span>
<span id="cb21-9"><a href="" aria-hidden="true" tabindex="-1"></a>      lab <span class="ot">&lt;-</span> <span class="fu">str_remove_all</span>(x, <span class="st">"KEGG_"</span>)</span>
<span id="cb21-10"><a href="" aria-hidden="true" tabindex="-1"></a>      <span class="fu">plotEnrichment</span>(kegg_by_gs[[x]], ranks_treatM) <span class="sc">+</span> <span class="fu">ggtitle</span>(lab)</span>
<span id="cb21-11"><a href="" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb21-12"><a href="" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb21-13"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wrap_plots</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

</section>
<section id="gene-set-enrichment-analysis-6-output" class="slide level2 output-location-slide"><h2>Gene Set Enrichment Analysis</h2><div class="cell output-location-slide" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="enrichment_slides_files/figure-revealjs/all-dir-barcode-1.png" class="quarto-figure quarto-figure-center" width="768"></p>
</figure>
</div>
</div>
</div></section><section id="gene-set-enrichment-analysis-7" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<div style="font-size: 90%;">
<ul>
<li>A quick summary of the results</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="" aria-hidden="true" tabindex="-1"></a>id2symbol <span class="ot">&lt;-</span> kegg_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb22-2"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(ensembl_gene, gene_symbol) <span class="sc">%&gt;%</span> </span>
<span id="cb22-3"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_symbol =</span> <span class="fu">setNames</span>(gene_symbol, ensembl_gene)) <span class="sc">%&gt;%</span> </span>
<span id="cb22-4"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(gene_symbol)</span>
<span id="cb22-5"><a href="" aria-hidden="true" tabindex="-1"></a>gsea_treatM <span class="sc">%&gt;%</span> </span>
<span id="cb22-6"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(pval) <span class="sc">%&gt;%</span> </span>
<span id="cb22-7"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(padj <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb22-8"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">leSize =</span> <span class="fu">map_int</span>(leadingEdge, length)) <span class="sc">%&gt;%</span> </span>
<span id="cb22-9"><a href="" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(pathway, NES, <span class="at">FDR =</span> padj, size, leSize, leadingEdge) <span class="sc">%&gt;%</span> </span>
<span id="cb22-10"><a href="" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb22-11"><a href="" aria-hidden="true" tabindex="-1"></a>    <span class="at">leadingEdge =</span> <span class="fu">lapply</span>(leadingEdge, \(x) id2symbol[x])</span>
<span id="cb22-12"><a href="" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                pathway       NES          FDR  size leSize
                                 &lt;char&gt;     &lt;num&gt;        &lt;num&gt; &lt;int&gt;  &lt;int&gt;
1:                        KEGG_LYSOSOME  1.883812 2.216231e-05   103     47
2:        KEGG_ECM_RECEPTOR_INTERACTION  1.749132 6.322486e-03    65     31
3:                  KEGG_FOCAL_ADHESION  1.621915 6.322486e-03   162     48
4: KEGG_DRUG_METABOLISM_CYTOCHROME_P450 -1.755175 4.366307e-02    21      9
                                leadingEdge
                                     &lt;list&gt;
1:     CTSH,CTSA,AP1S3,LIPA,LAMP3,FUCA1,...
2: THBS2,LAMA3,TNC,COL4A4,ITGB4,COL11A1,...
3:    MYL9,THBS2,LAMA3,TNC,COL4A4,PRKCB,...
4:   CYP2E1,MAOB,GSTM1,MGST2,MGST3,FMO4,...</code></pre>
</div>
</div>
</div>
<aside class="notes">
<ul>
<li>ECM_RECEPTOR_INTERACTION and FOCAL_ADHESION seem to be capturing a similar gene set</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="gene-set-enrichment-analysis-8" class="slide level2">
<h2>Gene Set Enrichment Analysis</h2>
<ul>
<li>This approach has assumed that clustering by direction is biologically relevant</li>
<li>Implies that up-regulated genes <span class="math inline">\(\implies\)</span> pathway activation</li>
<li>Ranks can also be prepared using significance only (<span class="math inline">\(-\log_{10}p\)</span>)
<ul>
<li>Effectively detects <em>perturbed</em> gene sets with genes at both ends</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="" aria-hidden="true" tabindex="-1"></a><span class="fu">fgsea</span>(kegg_by_gs, ranks_by_sig, <span class="at">scoreType =</span> <span class="st">"pos"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<div class="quarto-auto-generated-content">
<p><img src="assets/bodl_logo_white_background.jpg" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-fgsea2019" class="csl-entry" role="listitem">
Korotkevich, Gennady, Vladimir Sukhov, and Alexey Sergushichev. 2019. <span>“Fast Gene Set Enrichment Analysis.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/060012">https://doi.org/10.1101/060012</a>.
</div>
<div id="ref-Subramanian2005-lx" class="csl-entry" role="listitem">
Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005. <span>“Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.”</span> <em>Proc. Natl. Acad. Sci. U. S. A.</em> 102 (43): 15545–50.
</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="enrichment_slides_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="enrichment_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="enrichment_slides_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="enrichment_slides_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="enrichment_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="enrichment_slides_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="enrichment_slides_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="enrichment_slides_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="enrichment_slides_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="enrichment_slides_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    <script type="text/javascript">
      Reveal.on('ready', event => {
        if (event.indexh === 0) {
          document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
        }
      });
      Reveal.addEventListener('slidechanged', (event) => {
        if (event.indexh === 0) {
          Reveal.configure({ slideNumber: null });
          document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
        }
        if (event.indexh === 1) {
          Reveal.configure({ slideNumber: 'c' });
          document.querySelector("div.has-logo > img.slide-logo").style.display = null;
        }
      });
    </script>
    

</body></html>