<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dr Stevie Pederson">
<meta name="dcterms.date" content="2024-07-11">

<title>The Bioconductor Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="intro_bioc_slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="intro_bioc_slides_files/libs/quarto-html/quarto.js"></script>
<script src="intro_bioc_slides_files/libs/quarto-html/popper.min.js"></script>
<script src="intro_bioc_slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="intro_bioc_slides_files/libs/quarto-html/anchor.min.js"></script>
<link href="intro_bioc_slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="intro_bioc_slides_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="intro_bioc_slides_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="intro_bioc_slides_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="intro_bioc_slides_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  
  <script>
    (function () {
      var script = document.createElement("script");
      script.type = "text/javascript";
      script.src  = "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js";
      document.getElementsByTagName("head")[0].appendChild(script);
    })();
  </script>


<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="bodl.scss">
<link rel="stylesheet" href="extra.css">
</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#the-bioconductor-project" id="toc-the-bioconductor-project" class="nav-link active" data-scroll-target="#the-bioconductor-project">The Bioconductor Project</a></li>
  <li><a href="#object-classes" id="toc-object-classes" class="nav-link" data-scroll-target="#object-classes">Object Classes</a></li>
  <li><a href="#variations-on-data.frame-objects" id="toc-variations-on-data.frame-objects" class="nav-link" data-scroll-target="#variations-on-data.frame-objects">Variations on <code>data.frame</code> Objects</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="intro_bioc_slides.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">The Bioconductor Project</h1>
<p class="subtitle lead">RAdelaide 2024</p>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Dr Stevie Pederson </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            </p><div class="line-block">Black Ochre Data Labs<br>
            Telethon Kids Institute</div>
          <p></p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 11, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="the-bioconductor-project" class="level1" data-background-color="#3c3c44" data-background-image="https://raw.githubusercontent.com/Bioconductor/BiocStickers/devel/Bioconductor/Bioconductor5.png" data-background-size="30%" data-background-opacity="0.4" data-background-position="80% 50%">
<h1 data-background-color="#3c3c44" data-background-image="https://raw.githubusercontent.com/Bioconductor/BiocStickers/devel/Bioconductor/Bioconductor5.png" data-background-size="30%" data-background-opacity="0.4" data-background-position="80% 50%">The Bioconductor Project</h1>
<section id="todays-outline" class="level2">
<h2 class="anchored" data-anchor-id="todays-outline">Today’s Outline</h2>
<ul>
<li>Not just to show how to perform an RNA-Seq analysis</li>
<li>Understanding common file types</li>
<li>Understanding Bioconductor objects, classes &amp; methods
<ul>
<li>Has relevance beyond Bioconductor (e.g.&nbsp;map visualisation)</li>
</ul></li>
<li>Key concepts &amp; resources</li>
</ul>
<p>. . .</p>
<ul>
<li>Please ask questions!!! It’s the advantage of being here…</li>
</ul>
</section>
<section id="cran" class="level2">
<h2 class="anchored" data-anchor-id="cran">CRAN</h2>
<ul>
<li>A Package is a collection of functions</li>
<li>Associated with a given task/analysis/data-type</li>
<li>The main repository is “<strong>The Comprehensive R Archive Network</strong>” aka <a href="https://cran.r-project.org/">CRAN</a></li>
</ul>
<p>. . .</p>
<ul>
<li>We install packages using <code>Tools &gt; Install Packages...</code>
<ul>
<li>Or by typing <code>install.packages("pkg_name")</code></li>
</ul></li>
<li>Will only install packages from CRAN</li>
</ul>
</section>
<section id="cran-1" class="level2">
<h2 class="anchored" data-anchor-id="cran-1">CRAN</h2>
<ul>
<li>Packages on CRAN cover everything &amp; anything
<ul>
<li>GIS-based spatial data, stock trading, data visualisation etc</li>
</ul></li>
<li>Some packages are biological in focus, e.g.&nbsp;<code>Seurat</code> for scRNA</li>
</ul>
<p>. . .</p>
<ul>
<li>Submission involves passing technical checks
<ul>
<li>Tests across OSX, Windows &amp; Linux</li>
<li>Correct directory/file structures</li>
</ul></li>
</ul>
<p>. . .</p>
<ul>
<li>No checks for statistics that are methodologically sound
<ul>
<li>Usually publications associated <span class="math inline">\(\implies\)</span> peer-review</li>
</ul></li>
</ul>
</section>
<section id="packages" class="level2">
<h2 class="anchored" data-anchor-id="packages">Packages</h2>
<ul>
<li>Many packages are written by the authors for the authors</li>
<li>Decide to make them public in case others find them helpful</li>
</ul>
<p>. . .</p>
<ul>
<li>Both Chris &amp; I use <code>ngsReports</code> nearly every day (still…)</li>
</ul>
<p><img src="assets/ngsreports_pub.png" class="img-fluid"></p>
</section>
<section id="bioconductor" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor">Bioconductor</h2>
<ul>
<li>Bioconductor hosts packages focussed on biological research
<ul>
<li>www.bioconductor.org</li>
<li>Currently &gt; 2,400 packages</li>
</ul></li>
<li>Created by Robert Gentleman (one of the “R”s in <code>R</code>) in 2001</li>
</ul>
<p>. . .</p>
<ul>
<li>Packages checked for programming consistency<br><span class="math inline">\(\implies\)</span> not methodology
<ul>
<li>Expected to integrate with other Bioconductor packages</li>
</ul></li>
<li>Packages tested nightly on OSX, Windows, Debian Linux + Arch Linux</li>
<li>All packages require a vignette explaining how to use the package
<ul>
<li>Checked for clarity/helpfulness during package submission</li>
</ul></li>
</ul>
</section>
<section id="bioconductor-1" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-1">Bioconductor</h2>
<ul>
<li>Essentially represents a <em>community</em> of developers &amp; users</li>
<li>Strongly encourages &amp; supports diversity
<ul>
<li>No longer on Twitter</li>
</ul></li>
<li>Has a <a href="https://bioconductor.org/about/code-of-conduct/">Code of Conduct</a></li>
<li>Consciously supporting those from emerging countries
<ul>
<li>Actively trying to translate resources to other languages</li>
</ul></li>
<li>Also tries to find ways to engage with support developers &amp; users
<ul>
<li>support.bioconductor.org &amp; slack.bioconductor.org</li>
</ul></li>
</ul>
</section>
<section id="bioconductor-structure" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-structure">Bioconductor Structure</h2>
<div style="font-size: 95%;">
<ul>
<li><a href="https://bioconductor.org/about/core-team/">Core Team</a>
<ul>
<li>All employed by Bioconductor</li>
<li>Primarily grant funded (NIH, NSF, CZI)</li>
</ul></li>
<li><a href="https://bioconductor.org/about/scientific-advisory-board/">Scientific Advisory Board (SAB)</a>
<ul>
<li>Meet Annually.</li>
<li>External and Internal leaders who act as project advisors.</li>
</ul></li>
<li><a href="https://bioconductor.org/about/technical-advisory-board/">Technical Advisory Board (TAB)</a>
<ul>
<li>Meet monthly</li>
<li>Consider technical aspects of core infastructure and scientific direction of the project.</li>
<li>15 members, 3 year term. Annual, open elections to rotate members.</li>
</ul></li>
</ul>
</div>
</section>
<section id="bioconductor-structure-1" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-structure-1">Bioconductor Structure</h2>
<ul>
<li><a href="https://bioconductor.org/about/community-advisory-board/">Community Advisory Board (CAB)</a>
<ul>
<li>Meet monthly</li>
<li>Consider community outreach, events, education and training.</li>
<li>15 members, 3 year term. Annual, open elections to rotate members.</li>
</ul></li>
<li><a href="https://workinggroups.bioconductor.org">Multiple Working Groups</a>
<ul>
<li><a href="https://www.bioconductor.org/about/code-of-conduct/">Code of Conduct committee</a></li>
<li>Conference Committees, Training &amp; Education etc</li>
</ul></li>
</ul>
</section>
<section id="bioconductor-packages" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-packages">Bioconductor Packages</h2>
<ul>
<li><code>R</code> generally has bi-annual releases
<ul>
<li>With patch-fixes as needed</li>
<li>Latest release 4.4.1 (14<sup>th</sup> Jun, 2024)</li>
</ul></li>
<li>CRAN packages continually update</li>
</ul>
<p>. . .</p>
<ul>
<li>Bioconductor has two releases per year</li>
<li>Tested to a set R version (or later)</li>
<li>Bug-fixes also released as needed
<ul>
<li>Generally new features added to packages at Bioc releases</li>
</ul></li>
<li>Latest Bioconductor release is 3.19 (May 1<sup>st</sup> 2024)
<ul>
<li>Was tested on R v4.4.0</li>
</ul></li>
</ul>
</section>
<section id="bioconductor-packages-1" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-packages-1">Bioconductor Packages</h2>
<ul>
<li><code>BiocManager()</code> is a CRAN package</li>
<li>Enables installation from CRAN <em>and</em> Bioconductor
<ul>
<li>Also handles package installation from <code>github</code></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">"pkg1"</span>, <span class="st">"user/pkg2"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="bioconductor-packages-2" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-packages-2">Bioconductor Packages</h2>
<ul>
<li>Many packages for specific analyses
<ul>
<li><code>DESeq2</code> &amp; <code>edgeR</code> for bulk RNA-Seq Analysis</li>
<li><code>DiffBind</code> &amp; <code>extraChIPs</code> for ChIP-Seq Analysis</li>
<li><code>fgsea</code> for GSEA within <code>R</code></li>
</ul></li>
</ul>
<p>. . .</p>
<ul>
<li>Also multiple packages define object classes &amp; general methods
<ul>
<li>e.g.&nbsp;<code>GenomicRanges</code> for working with <code>GRanges</code> objects</li>
<li>Is a foundational class many other packages build on</li>
</ul></li>
<li>New packages are expected to use existing classes where possible</li>
</ul>
</section>
<section id="bioconductor-packages-3" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-packages-3">Bioconductor Packages</h2>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="https://carpentries-incubator.github.io/bioc-project/fig/bioc-sequencing-ecosystem.svg" class="img-fluid figure-img"></p>
<figcaption>Taken from https://carpentries-incubator.github.io/bioc-project/02-introduction-to-bioconductor.html</figcaption>
</figure>
</div>
</section>
<section id="bioconductor-packages-4" class="level2">
<h2 class="anchored" data-anchor-id="bioconductor-packages-4">Bioconductor Packages</h2>
<ul>
<li>The output from <code>body(function_name)</code> always has comments removed</li>
<li>All Bioconductor packages have the source code available
<ul>
<li>https://code.bioconductor.org</li>
</ul></li>
<li>Code is always written as a set of R Scripts</li>
<li>Inside the <code>R</code> directory
<ul>
<li>Will retain comments &amp; formatting from the authors</li>
<li>Sometimes more helpful…</li>
</ul></li>
</ul>
</section>
<section id="biocviews" class="level2">
<h2 class="anchored" data-anchor-id="biocviews">BiocViews</h2>
<ul>
<li><a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">BiocViews</a> provides on overview of all Bioconductor packages</li>
<li>Can be very helpful when looking for a resource</li>
<li>Most packages are software (for analysis)</li>
<li>Also annotation packages:
<ul>
<li>Genome sequences, gene to GO mappings etc</li>
</ul></li>
<li>Experimental data for demonstrating workflows</li>
<li>Workflow packages are slowly growing</li>
</ul>
<p>. . .</p>
<p><code>browseVignettes()</code></p>
</section>
</section>
<section id="object-classes" class="level1" data-background-color="#3c3c44">
<h1 data-background-color="#3c3c44">Object Classes</h1>
<section id="object-classes-1" class="level2">
<h2 class="anchored" data-anchor-id="object-classes-1">Object Classes</h2>
<p>R has two common types of objects</p>
<ul>
<li>Built on top of (and including) vectors, lists etc.</li>
<li><code>S3</code> are very common &amp; old (1970s)
<ul>
<li>Usually list-type objects e.g.&nbsp;results from <code>lm()</code> or <code>t.test()</code></li>
</ul></li>
<li><code>S4</code> introduced in ’90s
<ul>
<li>Focus on Object-Oriented Programming (OOP)<br>
</li>
</ul></li>
<li>Other less common classes (e.g.&nbsp;<code>R6</code> etc)</li>
</ul>
</section>
<section id="objects-and-methods" class="level2">
<h2 class="anchored" data-anchor-id="objects-and-methods">Objects and Methods</h2>
<ul>
<li>Functions can be written to handle different types of input data</li>
<li>Figuring out which version of the function to use<br><span class="math inline">\(\implies\)</span><em>method dispatch</em></li>
</ul>
<p>. . .</p>
<ul>
<li>A good example is the function <code>summary()</code>
<ul>
<li>Will return different results for a <code>vector</code> or <code>data.frame</code></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(letters)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mtcars)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>. . .</p>
<p><em>How does <code>summary()</code> know what to do for different data structures?</em></p>
</section>
<section id="objects-and-methods-1" class="level2">
<h2 class="anchored" data-anchor-id="objects-and-methods-1">Objects and Methods</h2>
<ul>
<li>If we try to look at the code used in <code>summary()</code> it’s a bit odd</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">body</span>(summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>UseMethod("summary")</code></pre>
</div>
</div>
<p>. . .</p>
<ul>
<li><code>summary()</code> uses different <strong>methods</strong> depending on the object class</li>
<li>Sometimes they’re hidden <span class="fragment">(I don’t know why…)</span></li>
</ul>
</section>
<section id="objects-and-methods-2" class="level2">
<h2 class="anchored" data-anchor-id="objects-and-methods-2">Objects and Methods</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">methods</span>(summary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] summary.aov                         summary.aovlist*                   
 [3] summary.aspell*                     summary.check_packages_in_dir*     
 [5] summary.connection                  summary.data.frame                 
 [7] summary.Date                        summary.default                    
 [9] summary.ecdf*                       summary.factor                     
[11] summary.glm                         summary.infl*                      
[13] summary.lm                          summary.loess*                     
[15] summary.manova                      summary.matrix                     
[17] summary.mlm*                        summary.nls*                       
[19] summary.packageStatus*              summary.POSIXct                    
[21] summary.POSIXlt                     summary.ppr*                       
[23] summary.prcomp*                     summary.princomp*                  
[25] summary.proc_time                   summary.rlang_error*               
[27] summary.rlang_message*              summary.rlang_trace*               
[29] summary.rlang_warning*              summary.rlang:::list_of_conditions*
[31] summary.srcfile                     summary.srcref                     
[33] summary.stepfun                     summary.stl*                       
[35] summary.table                       summary.tukeysmooth*               
[37] summary.warnings                   
see '?methods' for accessing help and source code</code></pre>
</div>
</div>
<div class="notes">
<p>The class is given after the dot Those marked with an asterisk are hidden</p>
</div>
</section>
<section id="objects-and-methods-3" class="level2">
<h2 class="anchored" data-anchor-id="objects-and-methods-3">Objects and Methods</h2>
<ul>
<li><code>summary.data.frame()</code> <br><span class="math inline">\(\implies\)</span> used when <code>summary()</code> is called on a <code>data.frame</code></li>
<li><code>summary.lm()</code> <br><span class="math inline">\(\implies\)</span> for an object produced by <code>lm()</code></li>
</ul>
<p>. . .</p>
<ul>
<li>If no method is written for a class <span class="math inline">\(\implies\)</span> <code>summary.default()</code></li>
<li>Look inside this using <code>body(summary.default)</code>
<ul>
<li>The last couple of lines were the output from <code>summary(letters)</code></li>
</ul></li>
</ul>
</section>
<section id="objects-and-methods-4" class="level2">
<h2 class="anchored" data-anchor-id="objects-and-methods-4">Objects and Methods</h2>
<ul>
<li>Can also see what methods exist for a given class</li>
<li>Before loading any packages ~56 methods exist for a data.frame</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">methods</span>(<span class="at">class =</span> <span class="st">"data.frame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] [             [[            [[&lt;-          [&lt;-           $&lt;-          
 [6] aggregate     anyDuplicated anyNA         as.data.frame as.list      
[11] as.matrix     as.vector     by            cbind         coerce       
[16] dim           dimnames      dimnames&lt;-    droplevels    duplicated   
[21] edit          format        formula       head          initialize   
[26] is.na         Math          merge         na.exclude    na.omit      
[31] Ops           plot          print         prompt        rbind        
[36] row.names     row.names&lt;-   rowsum        show          slotsFromS3  
[41] sort_by       split         split&lt;-       stack         str          
[46] subset        summary       Summary       t             tail         
[51] transform     type.convert  unique        unstack       within       
[56] xtfrm        
see '?methods' for accessing help and source code</code></pre>
</div>
</div>
</section>
<section id="objects-and-methods-5" class="level2">
<h2 class="anchored" data-anchor-id="objects-and-methods-5">Objects and Methods</h2>
<ul>
<li>Loading a new package may introduce new methods</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">methods</span>(<span class="at">class =</span> <span class="st">"data.frame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>. . .</p>
<ul>
<li>Now we have ~170 methods for a data.frame</li>
</ul>
</section>
<section id="objects-and-methods-6" class="level2">
<h2 class="anchored" data-anchor-id="objects-and-methods-6">Objects and Methods</h2>
<ul>
<li>Most classes have a <code>print()</code> method</li>
<li>Determines what to print to the screen when calling an object</li>
<li>Most common use case for me is <code>print(my_tbl, n = 20)</code>
<ul>
<li>Can use to override the default number of rows printed</li>
<li>Calls <code>print.tbl</code> (which is hidden)</li>
</ul></li>
</ul>
</section>
<section id="s3-objects" class="level2">
<h2 class="anchored" data-anchor-id="s3-objects">S3 Objects</h2>
<ul>
<li>Everything we’ve just seen applies to <code>S3</code> objects</li>
<li>Very common class type (<code>data.frame</code>, <code>list</code>, <code>htest</code>, <code>lm</code> etc)</li>
</ul>
<p>. . .</p>
<ul>
<li>Sometimes classes have an explicit hierarchy
<ul>
<li>Best shown using <code>is()</code> instead of <code>class()</code></li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(band_members)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "tbl_df"     "tbl"        "data.frame" "list"       "oldClass"  
[6] "vector"    </code></pre>
</div>
</div>
<p>. . .</p>
<ul>
<li><code>R</code> looks for <code>print.tbl_df()</code> <span class="math inline">\(\rightarrow\)</span> <code>print.tbl()</code> <span class="math inline">\(\rightarrow\)</span> <code>print.data.frame()</code> etc</li>
<li>Will use the first one found</li>
<li>If none found <span class="math inline">\(\implies\)</span> <code>print.default()</code></li>
</ul>
</section>
<section id="s4-objects" class="level2">
<h2 class="anchored" data-anchor-id="s4-objects">S4 Objects</h2>
<p>Many Bioconductor Packages define <code>S4</code> objects</p>
<ul>
<li>Very strict controls on data structure</li>
<li>Can be frustrating at first</li>
<li>Use the <code>@</code> symbol for “slots” as well as <code>$</code> for list elements
<ul>
<li>Slots are strictly defined components</li>
</ul></li>
<li>Methods are also strictly defined</li>
</ul>
</section>
<section id="s4-objects-1" class="level2">
<h2 class="anchored" data-anchor-id="s4-objects-1">S4 Objects</h2>
<ul>
<li>Can be a little more challenging to interact with the <code>tidyverse</code></li>
<li>Bioconductor pre-dates the <code>tidyverse</code> by &gt; 10 years</li>
<li><code>tidyomics</code> is an active area of Bioconductor development
<ul>
<li>Led by Stefano Mangiola from SAiGENCI</li>
</ul></li>
</ul>
<p>. . .</p>
<p><img src="assets/tidyomics.png" class="absolute img-fluid" data-left="0" data-bottom="0" width="740"></p>
</section>
<section id="s4-objects-2" class="level2">
<h2 class="anchored" data-anchor-id="s4-objects-2">S4 Objects</h2>
<div style="font-size: 95%;">
<ul>
<li>Some packages use <code>S4</code> implementations of <code>S3</code> objects
<ul>
<li><code>data.frame</code> (S3) Vs <code>DataFrame</code> (S4)</li>
<li><code>list</code> (S3) Vs <code>List</code> (S4)</li>
<li><code>vector</code> (S3) Vs <code>Vector</code> (S4)</li>
<li><code>rle</code> (S3) Vs <code>Rle</code> (S4)</li>
</ul></li>
</ul>
<div class="incremental">
<ul class="incremental">
<li>Many are written for memory efficiency</li>
<li>Look and behave similarly, but can ocassionally trip you over
<ul class="incremental">
<li>Object may require a <code>DataFrame</code> and you have a <code>data.frame</code></li>
<li>Coercion is usually relatively simple between base-level classes</li>
</ul></li>
</ul>
</div>
</div>
<div class="notes">
<p>Many S4 objects &amp; methods were developed in the days when compute resources were limited</p>
</div>
</section>
<section id="rle-vectors" class="level2">
<h2 class="anchored" data-anchor-id="rle-vectors">Rle Vectors</h2>
<ul>
<li>These are Run-Length Encoded vectors</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(S4Vectors)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"X"</span>, <span class="dv">10</span>), <span class="fu">rep</span>(<span class="st">"Y"</span>, <span class="dv">5</span>))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>test</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "X" "X" "X" "X" "X" "X" "X" "X" "X" "X" "Y" "Y" "Y" "Y" "Y"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Rle</span>(test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>character-Rle of length 15 with 2 runs
  Lengths:  10   5
  Values : "X" "Y"</code></pre>
</div>
</div>
<ul>
<li>Can encode millions of chromosomes with minimal memory
<ul>
<li>Sorting can help keep memory useage down</li>
</ul></li>
</ul>
</section>
</section>
<section id="variations-on-data.frame-objects" class="level1" data-background-color="#3c3c44">
<h1 data-background-color="#3c3c44">Variations on <code>data.frame</code> Objects</h1>
<section id="data.frame-objects" class="level2">
<h2 class="anchored" data-anchor-id="data.frame-objects"><code>data.frame</code> Objects</h2>
<ul>
<li><code>data.frame</code>
<ol type="1">
<li>Can set <code>rownames</code></li>
<li>Dumps all data to your screen</li>
<li>Cannot have column names with spaces</li>
</ol></li>
</ul>
<p>. . .</p>
<ul>
<li><code>tibble</code> aka <code>tbl_df</code>
<ol type="1">
<li><code>rownames</code> are always <code>1:nrow(df)</code></li>
<li>Prints a summary with rownames hidden</li>
<li>Column names with spaces permitted</li>
<li>My preferred <code>data.frame</code> type</li>
</ol></li>
</ul>
</section>
<section id="dataframe-objects" class="level2">
<h2 class="anchored" data-anchor-id="dataframe-objects"><code>DataFrame</code> objects</h2>
<div class="sourceCode" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>?DataFrame</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>An <code>S4</code> version
<ul>
<li>Doesn’t work with the <code>tidyverse</code></li>
<li>Still missing from <code>tidyomics</code></li>
</ul></li>
</ul>
<div class="fragment">
<ul>
<li>Until 2021 couldn’t coerce to a <code>tibble</code> directly</li>
<li>I hated that so wrote <code>extraChIPs::as_tibble()</code>
<ul>
<li>Enables passing <code>S4</code> objects to <code>ggplot()</code></li>
<li>Not as straight-forward as you might think</li>
<li>Please test &amp; find any bugs I haven’t found yet</li>
</ul></li>
</ul>
</div>
</section>
<section id="dataframe-objects-1" class="level2">
<h2 class="anchored" data-anchor-id="dataframe-objects-1"><code>DataFrame</code> objects</h2>
<ul>
<li>S3 Methods from <code>dplyr</code> will not work on <code>DataFrame</code> objects</li>
<li>Some equivalents exist (most pre-date the <code>tidyverse</code>)
<ul>
<li><code>subset()</code> pre-dates <code>dplyr::filter()</code></li>
<li><code>rbind()</code> and <code>combineRows()</code> <span class="math inline">\(\implies\)</span> <code>bind_rows()</code></li>
<li><code>cbind()</code>, <code>combineCols()</code> and <code>merge()</code> <span class="math inline">\(\implies\)</span> <code>joins</code></li>
<li><code>sort()</code> <span class="math inline">\(\implies\)</span> <code>arrange()</code></li>
<li><code>unique()</code> <span class="math inline">\(\implies\)</span> <code>distinct()</code></li>
</ul></li>
<li>No simple equivalent for <code>mutate()</code>, <code>summarise()</code>, <code>across()</code>, <code>pivot_*()</code></li>
</ul>
</section>
<section id="dataframe-objects-2" class="level2">
<h2 class="anchored" data-anchor-id="dataframe-objects-2"><code>DataFrame</code> objects</h2>
<ul>
<li>Can have columns of lists (so can <code>tbl_df</code> objects)
<ul>
<li>e.g.&nbsp;a <code>CharacterList()</code> from <code>IRanges</code></li>
<li><code>S4</code> lists can be <em>typed</em> <span class="math inline">\(\implies\)</span> memory efficiency</li>
<li><code>List</code> objects can exist in a compressed form <span class="math inline">\(\implies\)</span> memory efficiency</li>
</ul></li>
<li><code>DataFrame</code> objects can have <code>S4</code> objects as columns
<ul>
<li><code>S3</code> data frames <em>cannot</em></li>
</ul></li>
</ul>
<div class="notes">
<p>By typing a list we only need to record the type once, instead of once for each element. Can make a big difference with large objects</p>
</div>
</section>
<section id="dataframe-objects-3" class="level2">
<h2 class="anchored" data-anchor-id="dataframe-objects-3"><code>DataFrame</code> objects</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(IRanges)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>transcripts <span class="ot">&lt;-</span> <span class="fu">CharacterList</span>(</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"A3"</span>), <span class="fu">c</span>(<span class="st">"B1"</span>, <span class="st">"B2"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>transcripts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CharacterList of length 2
[[1]] A1 A2 A3
[[2]] B1 B2</code></pre>
</div>
</div>
<p>. . .</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>DF <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="at">Gene =</span> genes, <span class="at">Transcripts =</span> transcripts)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>DF</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 2 rows and 2 columns
         Gene     Transcripts
  &lt;character&gt; &lt;CharacterList&gt;
1           A        A1,A2,A3
2           B           B1,B2</code></pre>
</div>
</div>
<p>. . .</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(extraChIPs)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>(DF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  Gene  Transcripts
  &lt;chr&gt; &lt;list&gt;     
1 A     &lt;chr [3]&gt;  
2 B     &lt;chr [2]&gt;  </code></pre>
</div>
</div>
</section>
<section id="dataframe-objects-4" class="level2">
<h2 class="anchored" data-anchor-id="dataframe-objects-4"><code>DataFrame</code> objects</h2>
<ul>
<li>Metadata can be added to <code>DataFrame</code> objects</li>
<li>Must be a <code>list</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(DF) <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">details =</span> <span class="st">"Created for RAdelaide 2024"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(DF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'DFrame' [package "S4Vectors"] with 6 slots
  ..@ rownames       : NULL
  ..@ nrows          : int 2
  ..@ elementType    : chr "ANY"
  ..@ elementMetadata: NULL
  ..@ metadata       :List of 1
  .. ..$ details: chr "Created for RAdelaide 2024"
  ..@ listData       :List of 2
  .. ..$ Gene       : chr [1:2] "A" "B"
  .. ..$ Transcripts:Formal class 'CompressedCharacterList' [package "IRanges"] with 5 slots</code></pre>
</div>
</div>
<p>. . .</p>
<ul>
<li>Notice where rownames are and how the “columns” are stored</li>
</ul>
<div class="notes">
<p>Point out the <code>@</code> structure</p>
</div>
</section>
<section id="dataframe-objects-5" class="level2">
<h2 class="anchored" data-anchor-id="dataframe-objects-5"><code>DataFrame</code> objects</h2>
<ul>
<li>Also enable the addition of <em>column-specific metadata</em> <span class="math inline">\(\implies\)</span> <code>mcols()</code>
<ul>
<li>Used heavily in Bioconductor</li>
</ul></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(DF) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="at">meta =</span>  <span class="fu">c</span>(<span class="st">"Made-up genes"</span>, <span class="st">"Made-up transcripts"</span>))</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(DF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 2 rows and 1 column
                           meta
                    &lt;character&gt;
Gene              Made-up genes
Transcripts Made-up transcripts</code></pre>
</div>
</div>
<p>. . .</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(DF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'DFrame' [package "S4Vectors"] with 6 slots
  ..@ rownames       : NULL
  ..@ nrows          : int 2
  ..@ elementType    : chr "ANY"
  ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots
  ..@ metadata       :List of 1
  .. ..$ details: chr "Created for RAdelaide 2024"
  ..@ listData       :List of 2
  .. ..$ Gene       : chr [1:2] "A" "B"
  .. ..$ Transcripts:Formal class 'CompressedCharacterList' [package "IRanges"] with 5 slots</code></pre>
</div>
</div>
</section>
<section id="s4-object-structure" class="level2">
<h2 class="anchored" data-anchor-id="s4-object-structure"><code>S4</code> Object Structure</h2>
<ul>
<li><code>S4</code> objects have <em>slots</em> denoted with <code>@</code>
<ul>
<li>Subtly different to list <em>elements</em></li>
</ul></li>
<li>These are fixed and mandatory for every <code>S4</code> class
<ul>
<li>Can still be empty (<code>NULL</code>) objects</li>
<li>Can be <code>S3</code> or <code>S4</code> objects</li>
</ul></li>
<li>Enforces a strict structure with checks as objects formed
<ul>
<li>Saves time performing checks within functions</li>
</ul></li>
</ul>
</section>
<section id="s4-object-structure-1" class="level2">
<h2 class="anchored" data-anchor-id="s4-object-structure-1"><code>S4</code> Object Structure</h2>
<ul>
<li>We can’t <code>lapply</code> our way through these objects</li>
<li>Can access each slot using shortcuts <code>object@slotName</code>
<ul>
<li>More formally using <code>slot(object, "slotName")</code></li>
</ul></li>
</ul>
<div class="columns">
<div class="column">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>DF<span class="sc">@</span>listData</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$Gene
[1] "A" "B"

$Transcripts
CharacterList of length 2
[[1]] A1 A2 A3
[[2]] B1 B2</code></pre>
</div>
</div>
</div><div class="column">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">slot</span>(DF, <span class="st">"listData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$Gene
[1] "A" "B"

$Transcripts
CharacterList of length 2
[[1]] A1 A2 A3
[[2]] B1 B2</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="s4-object-structure-2" class="level2">
<h2 class="anchored" data-anchor-id="s4-object-structure-2"><code>S4</code> Object Structure</h2>
<ul>
<li>Slot names can be found using <code>slotNames(object)</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">slotNames</span>(DF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "rownames"        "nrows"           "elementType"     "elementMetadata"
[5] "metadata"        "listData"       </code></pre>
</div>
</div>
<p>. . .</p>
<ul>
<li>The full type description can also be found</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getSlots</span>(<span class="st">"DFrame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           rownames               nrows         elementType     elementMetadata 
"character_OR_NULL"           "integer"         "character" "DataFrame_OR_NULL" 
           metadata            listData 
             "list"              "list" </code></pre>
</div>
</div>
</section>
<section id="s4-methods" class="level2">
<h2 class="anchored" data-anchor-id="s4-methods"><code>S4</code> Methods</h2>
<ul>
<li><code>S3</code> method dispatch uses the <code>method.class</code> syntax</li>
<li><code>S4</code> is very different but has some similarities</li>
</ul>
<p>. . .</p>
<ul>
<li><code>S4</code> objects almost always have hierarchical classes
<ul>
<li>Increasingly common for <code>S3</code> objects</li>
<li>Each level <em>extends</em> the lower level class</li>
</ul></li>
</ul>
<p>. . .</p>
<ul>
<li>Methods are strictly defined by package authors
<ul>
<li>A <code>Generic</code> function must be defined for each method/class</li>
<li>The hierarchy is traversed until a method is found</li>
</ul></li>
</ul>
</section>
<section id="s4-methods-1" class="level2">
<h2 class="anchored" data-anchor-id="s4-methods-1"><code>S4</code> Methods</h2>
<ul>
<li>To check the class hierarchy of an object use <code>is()</code></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(DF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "DFrame"            "DataFrame"         "SimpleList"       
 [4] "RectangularData"   "List"              "DataFrame_OR_NULL"
 [7] "Vector"            "list_OR_List"      "Annotated"        
[10] "vector_OR_Vector" </code></pre>
</div>
</div>
<p>. . .</p>
<ul>
<li>Logical tests can be applied</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(DF, <span class="st">"DataFrame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(DF, <span class="st">"data.frame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<p>. . .</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">methods</span>(<span class="at">class =</span> <span class="st">"DataFrame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="s4-methods-2" class="level2">
<h2 class="anchored" data-anchor-id="s4-methods-2"><code>S4</code> Methods</h2>
<ul>
<li>The function <code>body()</code> will return <code>standardGeneric()</code>
<ul>
<li>Slightly different to <code>UseMethod()</code></li>
</ul></li>
</ul>
<p>. . .</p>
<ul>
<li>To show code within a function</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getMethod</span>(<span class="at">f =</span> <span class="st">"nrow"</span>, <span class="at">signature =</span> <span class="st">"DataFrame"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Method Definition:

function (x) 
x@nrows
&lt;bytecode: 0x55e8874d0dc8&gt;
&lt;environment: namespace:S4Vectors&gt;

Signatures:
        x          
target  "DataFrame"
defined "DataFrame"</code></pre>
</div>
</div>
</section>
<section id="recap" class="level2">
<h2 class="anchored" data-anchor-id="recap">Recap</h2>
<ul>
<li>The Bioconductor Project extends back to the early days of <code>R</code>
<ul>
<li>Also the early days of bioinformatics</li>
<li>Has a genuine community aspect . . .</li>
</ul></li>
<li><code>S4</code> object classes are common
<ul>
<li>Foundational classes enabling package inter-operability</li>
<li>Much less common in <code>CRAN</code> packages (spatial/GIS)</li>
</ul></li>
<li>Method dispatch is handled differently</li>
</ul>
<p>. . .</p>
<ul>
<li>Can play badly with the <code>tidyverse</code>
<ul>
<li>An area of active development</li>
</ul></li>
</ul>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<script type="text/javascript">
  Reveal.on('ready', event => {
    if (event.indexh === 0) {
      document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
    }
  });
  Reveal.addEventListener('slidechanged', (event) => {
    if (event.indexh === 0) {
      Reveal.configure({ slideNumber: null });
      document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
    }
    if (event.indexh === 1) {
      Reveal.configure({ slideNumber: 'c' });
      document.querySelector("div.has-logo > img.slide-logo").style.display = null;
    }
  });
</script>




</body></html>